{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "# importing libraries\n",
        "import zipfile\n",
        "import os\n"
      ],
      "metadata": {
        "id": "1kDdlSYZP5Vs"
      },
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install bcftools\n",
        "\n",
        "# for calculating ser\n",
        "!pip3 install --user whatshap\n",
        "\n",
        "# to prevent whatshap command not found\n",
        "os.environ['PATH'] += \":/root/.local/bin\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MvNRL35zQRnH",
        "outputId": "6262e00f-7d73-47e5-85e8-135e9f0f9153"
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "bcftools is already the newest version (1.13-1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.\n",
            "Requirement already satisfied: whatshap in /root/.local/lib/python3.10/site-packages (2.3)\n",
            "Requirement already satisfied: pysam>=0.18.0 in /root/.local/lib/python3.10/site-packages (from whatshap) (0.22.1)\n",
            "Requirement already satisfied: pyfaidx>=0.5.5.2 in /root/.local/lib/python3.10/site-packages (from whatshap) (0.8.1.3)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from whatshap) (3.4.2)\n",
            "Requirement already satisfied: biopython>=1.73 in /root/.local/lib/python3.10/site-packages (from whatshap) (1.84)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (from whatshap) (1.13.1)\n",
            "Requirement already satisfied: xopen>=1.2.0 in /root/.local/lib/python3.10/site-packages (from whatshap) (2.0.2)\n",
            "Requirement already satisfied: pulp>=2 in /root/.local/lib/python3.10/site-packages (from whatshap) (2.9.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython>=1.73->whatshap) (1.26.4)\n",
            "Requirement already satisfied: importlib-metadata in /usr/local/lib/python3.10/dist-packages (from pyfaidx>=0.5.5.2->whatshap) (8.5.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from pyfaidx>=0.5.5.2->whatshap) (24.2)\n",
            "Requirement already satisfied: isal>=1.6.1 in /root/.local/lib/python3.10/site-packages (from xopen>=1.2.0->whatshap) (1.7.1)\n",
            "Requirement already satisfied: zlib-ng>=0.4.1 in /root/.local/lib/python3.10/site-packages (from xopen>=1.2.0->whatshap) (0.5.1)\n",
            "Requirement already satisfied: zipp>=3.20 in /usr/local/lib/python3.10/dist-packages (from importlib-metadata->pyfaidx>=0.5.5.2->whatshap) (3.21.0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aWyCU4HrPLPf",
        "outputId": "3898d360-203c-491e-8bdb-d14386b424ec"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "folder = 'CG-HW3-Phasing'"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# read files\n",
        "with zipfile.ZipFile(f'/content/drive/My Drive/DataHW03.zip', 'r') as zip_ref:\n",
        "    zip_ref.extractall('/content/')"
      ],
      "metadata": {
        "id": "Xj-3SFjnQG8S"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Part 1\n",
        "unphased_data = f'/content/DataHW03/chr14.GRCh38.UnPhased.vcf.gz'\n",
        "\n",
        "# shows hearder columns\n",
        "!bcftools view -h {unphased_data}\n",
        "# shows number of header rows\n",
        "# (they are at the start of the file and contain metadata)\n",
        "!bcftools view -h {unphased_data} | wc -l\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TMOLcfEkRD-Q",
        "outputId": "b6a81abe-050c-426a-96b2-d1e77125b7e7"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "##fileformat=VCFv4.3\n",
            "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
            "##fileDate=27022019_15h52m43s\n",
            "##source=IGSRpipeline\n",
            "##reference=ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa\n",
            "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"UnPhased Genotype\">\n",
            "##contig=<ID=14>\n",
            "##INFO=<ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1)\">\n",
            "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Total number of alternate alleles in called genotypes\">\n",
            "##INFO=<ID=NS,Number=1,Type=Integer,Description=\"Number of samples with data\">\n",
            "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
            "##INFO=<ID=EAS_AF,Number=A,Type=Float,Description=\"Allele frequency in the EAS populations calculated from AC and AN, in the range (0,1)\">\n",
            "##INFO=<ID=EUR_AF,Number=A,Type=Float,Description=\"Allele frequency in the EUR populations calculated from AC and AN, in the range (0,1)\">\n",
            "##INFO=<ID=AFR_AF,Number=A,Type=Float,Description=\"Allele frequency in the AFR populations calculated from AC and AN, in the range (0,1)\">\n",
            "##INFO=<ID=AMR_AF,Number=A,Type=Float,Description=\"Allele frequency in the AMR populations calculated from AC and AN, in the range (0,1)\">\n",
            "##INFO=<ID=SAS_AF,Number=A,Type=Float,Description=\"Allele frequency in the SAS populations calculated from AC and AN, in the range (0,1)\">\n",
            "##INFO=<ID=VT,Number=.,Type=String,Description=\"indicates what type of variant the line represents\">\n",
            "##INFO=<ID=EX_TARGET,Number=0,Type=Flag,Description=\"indicates whether a variant is within the exon pull down target boundaries\">\n",
            "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth; some reads may have been filtered\">\n",
            "##bcftools_viewVersion=1.13+htslib-1.13+ds\n",
            "##bcftools_viewCommand=view -h /content/DataHW03/chr14.GRCh38.UnPhased.vcf.gz; Date=Sat Dec 28 22:29:01 2024\n",
            "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tHG00096\tHG00097\tHG00099\tHG00100\tHG00101\tHG00102\tHG00103\tHG00104\tHG00105\tHG00106\tHG00107\tHG00108\tHG00109\tHG00110\tHG00111\tHG00112\tHG00113\tHG00114\tHG00115\tHG00116\tHG00117\tHG00118\tHG00119\tHG00120\tHG00121\tHG00122\tHG00123\tHG00125\tHG00126\tHG00127\tHG00128\tHG00129\tHG00130\tHG00131\tHG00132\tHG00133\tHG00134\tHG00135\tHG00136\tHG00137\tHG00138\tHG00139\tHG00140\tHG00141\tHG00142\tHG00143\tHG00145\tHG00146\tHG00148\tHG00149\tHG00150\tHG00151\tHG00152\tHG00154\tHG00155\tHG00156\tHG00157\tHG00158\tHG00159\tHG00160\tHG00171\tHG00173\tHG00174\tHG00176\tHG00177\tHG00178\tHG00179\tHG00180\tHG00181\tHG00182\tHG00183\tHG00185\tHG00186\tHG00187\tHG00188\tHG00189\tHG00190\tHG00231\tHG00232\tHG00233\tHG00234\tHG00235\tHG00236\tHG00237\tHG00238\tHG00239\tHG00240\tHG00242\tHG00243\tHG00244\tHG00245\tHG00246\tHG00249\tHG00250\tHG00251\tHG00252\tHG00253\tHG00254\tHG00255\tHG00256\tHG00257\tHG00258\tHG00259\tHG00260\tHG00261\tHG00262\tHG00263\tHG00264\tHG00265\tHG00266\tHG00267\tHG00268\tHG00269\tHG00270\tHG00271\tHG00272\tHG00273\tHG00274\tHG00275\tHG00276\tHG00277\tHG00278\tHG00280\tHG00281\tHG00282\tHG00284\tHG00285\tHG00288\tHG00290\tHG00302\tHG00303\tHG00304\tHG00306\tHG00308\tHG00309\tHG00310\tHG00311\tHG00312\tHG00313\tHG00315\tHG00318\tHG00319\tHG00320\tHG00321\tHG00323\tHG00324\tHG00325\tHG00326\tHG00327\tHG00328\tHG00329\tHG00330\tHG00331\tHG00332\tHG00334\tHG00335\tHG00336\tHG00337\tHG00338\tHG00339\tHG00341\tHG00342\tHG00343\tHG00344\tHG00345\tHG00346\tHG00349\tHG00350\tHG00351\tHG00353\tHG00355\tHG00356\tHG00357\tHG00358\tHG00359\tHG00360\tHG00361\tHG00362\tHG00364\tHG00365\tHG00366\tHG00367\tHG00368\tHG00369\tHG00371\tHG00372\tHG00373\tHG00375\tHG00376\tHG00377\tHG00378\tHG00379\tHG00380\tHG00381\tHG00382\tHG00383\tHG00384\tHG00403\tHG00404\tHG00406\tHG00407\tHG00409\tHG00410\tHG00419\tHG00421\tHG00422\tHG00428\tHG00436\tHG00437\tHG00442\tHG00443\tHG00445\tHG00446\tHG00448\tHG00449\tHG00451\tHG00452\tHG00457\tHG00458\tHG00463\tHG00464\tHG00472\tHG00473\tHG00475\tHG00476\tHG00478\tHG00479\tHG00500\tHG00513\tHG00524\tHG00525\tHG00530\tHG00531\tHG00533\tHG00534\tHG00536\tHG00537\tHG00542\tHG00543\tHG00551\tHG00553\tHG00554\tHG00556\tHG00557\tHG00559\tHG00560\tHG00565\tHG00566\tHG00580\tHG00581\tHG00583\tHG00584\tHG00589\tHG00590\tHG00592\tHG00593\tHG00595\tHG00596\tHG00598\tHG00599\tHG00607\tHG00608\tHG00610\tHG00611\tHG00613\tHG00614\tHG00619\tHG00620\tHG00622\tHG00623\tHG00625\tHG00626\tHG00628\tHG00629\tHG00631\tHG00632\tHG00634\tHG00637\tHG00638\tHG00640\tHG00641\tHG00650\tHG00651\tHG00653\tHG00654\tHG00656\tHG00657\tHG00662\tHG00663\tHG00671\tHG00672\tHG00674\tHG00675\tHG00683\tHG00684\tHG00689\tHG00690\tHG00692\tHG00693\tHG00698\tHG00699\tHG00701\tHG00704\tHG00705\tHG00707\tHG00708\tHG00717\tHG00728\tHG00729\tHG00731\tHG00732\tHG00734\tHG00736\tHG00737\tHG00739\tHG00740\tHG00742\tHG00743\tHG00759\tHG00766\tHG00844\tHG00851\tHG00864\tHG00867\tHG00879\tHG00881\tHG00956\tHG00978\tHG00982\tHG01028\tHG01029\tHG01031\tHG01046\tHG01047\tHG01048\tHG01049\tHG01051\tHG01052\tHG01054\tHG01055\tHG01058\tHG01060\tHG01061\tHG01063\tHG01064\tHG01066\tHG01067\tHG01069\tHG01070\tHG01072\tHG01073\tHG01075\tHG01077\tHG01079\tHG01080\tHG01082\tHG01083\tHG01085\tHG01086\tHG01088\tHG01089\tHG01092\tHG01094\tHG01095\tHG01097\tHG01098\tHG01101\tHG01102\tHG01104\tHG01105\tHG01107\tHG01108\tHG01110\tHG01111\tHG01112\tHG01113\tHG01119\tHG01121\tHG01122\tHG01124\tHG01125\tHG01130\tHG01131\tHG01133\tHG01134\tHG01136\tHG01137\tHG01139\tHG01140\tHG01142\tHG01148\tHG01149\tHG01161\tHG01162\tHG01164\tHG01167\tHG01168\tHG01170\tHG01171\tHG01173\tHG01174\tHG01176\tHG01177\tHG01182\tHG01183\tHG01187\tHG01188\tHG01190\tHG01191\tHG01197\tHG01198\tHG01200\tHG01204\tHG01205\tHG01241\tHG01242\tHG01247\tHG01248\tHG01250\tHG01251\tHG01253\tHG01254\tHG01256\tHG01257\tHG01259\tHG01260\tHG01269\tHG01271\tHG01272\tHG01275\tHG01277\tHG01280\tHG01281\tHG01284\tHG01286\tHG01302\tHG01303\tHG01305\tHG01308\tHG01311\tHG01312\tHG01323\tHG01325\tHG01326\tHG01334\tHG01341\tHG01342\tHG01344\tHG01345\tHG01348\tHG01350\tHG01351\tHG01353\tHG01354\tHG01356\tHG01357\tHG01359\tHG01360\tHG01362\tHG01363\tHG01365\tHG01366\tHG01369\tHG01372\tHG01374\tHG01375\tHG01377\tHG01378\tHG01383\tHG01384\tHG01389\tHG01390\tHG01392\tHG01393\tHG01395\tHG01396\tHG01398\tHG01402\tHG01403\tHG01405\tHG01412\tHG01413\tHG01414\tHG01431\tHG01432\tHG01435\tHG01437\tHG01438\tHG01440\tHG01441\tHG01443\tHG01444\tHG01447\tHG01455\tHG01456\tHG01459\tHG01461\tHG01462\tHG01464\tHG01465\tHG01468\tHG01471\tHG01474\tHG01479\tHG01485\tHG01486\tHG01488\tHG01489\tHG01491\tHG01492\tHG01494\tHG01495\tHG01497\tHG01498\tHG01500\tHG01501\tHG01503\tHG01504\tHG01506\tHG01507\tHG01509\tHG01510\tHG01512\tHG01513\tHG01515\tHG01516\tHG01518\tHG01519\tHG01521\tHG01522\tHG01524\tHG01525\tHG01527\tHG01528\tHG01530\tHG01531\tHG01536\tHG01537\tHG01550\tHG01551\tHG01556\tHG01565\tHG01566\tHG01571\tHG01572\tHG01577\tHG01578\tHG01583\tHG01586\tHG01589\tHG01593\tHG01595\tHG01596\tHG01597\tHG01598\tHG01599\tHG01600\tHG01602\tHG01603\tHG01605\tHG01606\tHG01607\tHG01608\tHG01610\tHG01612\tHG01613\tHG01615\tHG01617\tHG01618\tHG01619\tHG01620\tHG01623\tHG01624\tHG01625\tHG01626\tHG01628\tHG01630\tHG01631\tHG01632\tHG01668\tHG01669\tHG01670\tHG01672\tHG01673\tHG01675\tHG01676\tHG01678\tHG01679\tHG01680\tHG01682\tHG01684\tHG01685\tHG01686\tHG01694\tHG01695\tHG01697\tHG01699\tHG01700\tHG01702\tHG01704\tHG01705\tHG01707\tHG01708\tHG01709\tHG01710\tHG01746\tHG01747\tHG01756\tHG01757\tHG01761\tHG01762\tHG01765\tHG01766\tHG01767\tHG01768\tHG01770\tHG01771\tHG01773\tHG01775\tHG01776\tHG01777\tHG01779\tHG01781\tHG01783\tHG01784\tHG01785\tHG01786\tHG01789\tHG01790\tHG01791\tHG01794\tHG01795\tHG01796\tHG01797\tHG01798\tHG01799\tHG01800\tHG01801\tHG01802\tHG01804\tHG01805\tHG01806\tHG01807\tHG01808\tHG01809\tHG01810\tHG01811\tHG01812\tHG01813\tHG01815\tHG01816\tHG01817\tHG01840\tHG01841\tHG01842\tHG01843\tHG01844\tHG01845\tHG01846\tHG01847\tHG01848\tHG01849\tHG01850\tHG01851\tHG01852\tHG01853\tHG01855\tHG01857\tHG01858\tHG01859\tHG01860\tHG01861\tHG01862\tHG01863\tHG01864\tHG01865\tHG01866\tHG01867\tHG01868\tHG01869\tHG01870\tHG01871\tHG01872\tHG01873\tHG01874\tHG01878\tHG01879\tHG01880\tHG01882\tHG01883\tHG01885\tHG01886\tHG01889\tHG01890\tHG01892\tHG01893\tHG01894\tHG01896\tHG01912\tHG01914\tHG01915\tHG01917\tHG01918\tHG01920\tHG01921\tHG01923\tHG01924\tHG01926\tHG01927\tHG01932\tHG01933\tHG01935\tHG01936\tHG01938\tHG01939\tHG01941\tHG01942\tHG01944\tHG01945\tHG01947\tHG01948\tHG01950\tHG01951\tHG01953\tHG01954\tHG01956\tHG01958\tHG01961\tHG01965\tHG01967\tHG01968\tHG01970\tHG01971\tHG01973\tHG01974\tHG01976\tHG01977\tHG01979\tHG01980\tHG01982\tHG01985\tHG01986\tHG01988\tHG01989\tHG01990\tHG01991\tHG01992\tHG01997\tHG02002\tHG02003\tHG02006\tHG02008\tHG02009\tHG02010\tHG02012\tHG02013\tHG02014\tHG02016\tHG02017\tHG02019\tHG02020\tHG02023\tHG02025\tHG02026\tHG02028\tHG02029\tHG02031\tHG02032\tHG02035\tHG02040\tHG02047\tHG02048\tHG02049\tHG02050\tHG02051\tHG02052\tHG02053\tHG02054\tHG02057\tHG02058\tHG02060\tHG02061\tHG02064\tHG02067\tHG02069\tHG02070\tHG02072\tHG02073\tHG02075\tHG02076\tHG02078\tHG02079\tHG02081\tHG02082\tHG02084\tHG02085\tHG02086\tHG02087\tHG02088\tHG02089\tHG02090\tHG02095\tHG02102\tHG02104\tHG02105\tHG02107\tHG02108\tHG02111\tHG02113\tHG02116\tHG02121\tHG02122\tHG02127\tHG02128\tHG02130\tHG02131\tHG02133\tHG02134\tHG02136\tHG02137\tHG02138\tHG02139\tHG02140\tHG02141\tHG02142\tHG02143\tHG02144\tHG02146\tHG02147\tHG02150\tHG02151\tHG02152\tHG02153\tHG02154\tHG02155\tHG02156\tHG02164\tHG02165\tHG02166\tHG02168\tHG02169\tHG02170\tHG02173\tHG02176\tHG02178\tHG02179\tHG02180\tHG02181\tHG02182\tHG02184\tHG02185\tHG02186\tHG02187\tHG02188\tHG02190\tHG02215\tHG02219\tHG02220\tHG02221\tHG02223\tHG02224\tHG02230\tHG02231\tHG02232\tHG02233\tHG02235\tHG02236\tHG02238\tHG02239\tHG02250\tHG02252\tHG02253\tHG02255\tHG02256\tHG02259\tHG02260\tHG02262\tHG02265\tHG02266\tHG02271\tHG02272\tHG02274\tHG02275\tHG02277\tHG02278\tHG02281\tHG02282\tHG02283\tHG02284\tHG02285\tHG02286\tHG02291\tHG02292\tHG02298\tHG02299\tHG02301\tHG02304\tHG02307\tHG02308\tHG02309\tHG02312\tHG02314\tHG02315\tHG02317\tHG02318\tHG02322\tHG02323\tHG02325\tHG02330\tHG02332\tHG02334\tHG02337\tHG02339\tHG02343\tHG02345\tHG02348\tHG02351\tHG02353\tHG02355\tHG02356\tHG02358\tHG02360\tHG02364\tHG02367\tHG02371\tHG02373\tHG02374\tHG02375\tHG02379\tHG02380\tHG02382\tHG02383\tHG02384\tHG02385\tHG02386\tHG02389\tHG02390\tHG02391\tHG02392\tHG02394\tHG02395\tHG02396\tHG02397\tHG02398\tHG02399\tHG02401\tHG02402\tHG02405\tHG02406\tHG02407\tHG02408\tHG02409\tHG02410\tHG02419\tHG02420\tHG02425\tHG02427\tHG02429\tHG02433\tHG02436\tHG02439\tHG02442\tHG02445\tHG02449\tHG02450\tHG02455\tHG02461\tHG02462\tHG02464\tHG02465\tHG02470\tHG02471\tHG02476\tHG02477\tHG02479\tHG02481\tHG02484\tHG02485\tHG02489\tHG02490\tHG02491\tHG02493\tHG02494\tHG02496\tHG02497\tHG02501\tHG02502\tHG02505\tHG02508\tHG02511\tHG02512\tHG02513\tHG02521\tHG02522\tHG02536\tHG02537\tHG02541\tHG02545\tHG02546\tHG02549\tHG02554\tHG02555\tHG02557\tHG02558\tHG02561\tHG02562\tHG02568\tHG02570\tHG02571\tHG02573\tHG02574\tHG02577\tHG02580\tHG02582\tHG02583\tHG02585\tHG02586\tHG02588\tHG02589\tHG02594\tHG02595\tHG02597\tHG02600\tHG02601\tHG02603\tHG02604\tHG02610\tHG02611\tHG02613\tHG02614\tHG02620\tHG02621\tHG02623\tHG02624\tHG02628\tHG02629\tHG02634\tHG02635\tHG02642\tHG02643\tHG02645\tHG02646\tHG02648\tHG02649\tHG02651\tHG02652\tHG02654\tHG02655\tHG02657\tHG02658\tHG02660\tHG02661\tHG02666\tHG02667\tHG02675\tHG02676\tHG02678\tHG02679\tHG02681\tHG02682\tHG02684\tHG02685\tHG02687\tHG02688\tHG02690\tHG02691\tHG02694\tHG02696\tHG02697\tHG02699\tHG02700\tHG02702\tHG02703\tHG02715\tHG02716\tHG02721\tHG02722\tHG02724\tHG02725\tHG02727\tHG02728\tHG02731\tHG02733\tHG02734\tHG02736\tHG02737\tHG02756\tHG02757\tHG02759\tHG02760\tHG02763\tHG02768\tHG02769\tHG02771\tHG02772\tHG02774\tHG02775\tHG02778\tHG02780\tHG02783\tHG02784\tHG02786\tHG02787\tHG02789\tHG02790\tHG02792\tHG02793\tHG02798\tHG02799\tHG02804\tHG02805\tHG02807\tHG02808\tHG02810\tHG02811\tHG02813\tHG02814\tHG02816\tHG02817\tHG02819\tHG02820\tHG02836\tHG02837\tHG02839\tHG02840\tHG02851\tHG02852\tHG02854\tHG02855\tHG02860\tHG02861\tHG02870\tHG02878\tHG02879\tHG02881\tHG02882\tHG02884\tHG02885\tHG02887\tHG02888\tHG02890\tHG02891\tHG02895\tHG02896\tHG02922\tHG02923\tHG02938\tHG02941\tHG02943\tHG02944\tHG02946\tHG02947\tHG02952\tHG02953\tHG02968\tHG02970\tHG02971\tHG02973\tHG02974\tHG02976\tHG02977\tHG02979\tHG02981\tHG02982\tHG02983\tHG03006\tHG03007\tHG03009\tHG03012\tHG03015\tHG03016\tHG03018\tHG03019\tHG03021\tHG03022\tHG03024\tHG03025\tHG03027\tHG03028\tHG03039\tHG03040\tHG03045\tHG03046\tHG03048\tHG03049\tHG03052\tHG03054\tHG03055\tHG03057\tHG03058\tHG03060\tHG03061\tHG03063\tHG03064\tHG03066\tHG03069\tHG03072\tHG03073\tHG03074\tHG03077\tHG03078\tHG03079\tHG03081\tHG03082\tHG03084\tHG03085\tHG03086\tHG03088\tHG03091\tHG03095\tHG03096\tHG03097\tHG03099\tHG03100\tHG03103\tHG03105\tHG03108\tHG03109\tHG03111\tHG03112\tHG03114\tHG03115\tHG03117\tHG03118\tHG03120\tHG03121\tHG03123\tHG03124\tHG03126\tHG03127\tHG03129\tHG03130\tHG03132\tHG03133\tHG03135\tHG03136\tHG03139\tHG03157\tHG03159\tHG03160\tHG03162\tHG03163\tHG03166\tHG03168\tHG03169\tHG03171\tHG03172\tHG03175\tHG03189\tHG03190\tHG03193\tHG03195\tHG03196\tHG03198\tHG03199\tHG03202\tHG03209\tHG03212\tHG03224\tHG03225\tHG03228\tHG03229\tHG03234\tHG03235\tHG03237\tHG03238\tHG03240\tHG03241\tHG03246\tHG03247\tHG03258\tHG03259\tHG03265\tHG03267\tHG03268\tHG03270\tHG03271\tHG03279\tHG03280\tHG03291\tHG03294\tHG03295\tHG03297\tHG03298\tHG03300\tHG03301\tHG03303\tHG03304\tHG03311\tHG03313\tHG03342\tHG03343\tHG03351\tHG03352\tHG03354\tHG03363\tHG03366\tHG03367\tHG03369\tHG03370\tHG03372\tHG03376\tHG03378\tHG03380\tHG03382\tHG03385\tHG03388\tHG03391\tHG03393\tHG03394\tHG03397\tHG03398\tHG03401\tHG03410\tHG03419\tHG03428\tHG03431\tHG03432\tHG03433\tHG03436\tHG03437\tHG03439\tHG03442\tHG03445\tHG03446\tHG03449\tHG03451\tHG03452\tHG03455\tHG03457\tHG03458\tHG03460\tHG03461\tHG03462\tHG03464\tHG03469\tHG03470\tHG03472\tHG03473\tHG03476\tHG03478\tHG03479\tHG03484\tHG03485\tHG03488\tHG03490\tHG03491\tHG03499\tHG03511\tHG03514\tHG03515\tHG03517\tHG03518\tHG03520\tHG03521\tHG03538\tHG03539\tHG03547\tHG03548\tHG03549\tHG03556\tHG03557\tHG03558\tHG03559\tHG03563\tHG03565\tHG03567\tHG03571\tHG03572\tHG03575\tHG03577\tHG03578\tHG03583\tHG03585\tHG03589\tHG03593\tHG03594\tHG03595\tHG03598\tHG03600\tHG03603\tHG03604\tHG03607\tHG03611\tHG03615\tHG03616\tHG03619\tHG03624\tHG03625\tHG03629\tHG03631\tHG03634\tHG03636\tHG03640\tHG03642\tHG03643\tHG03644\tHG03645\tHG03646\tHG03649\tHG03652\tHG03653\tHG03660\tHG03663\tHG03667\tHG03668\tHG03672\tHG03673\tHG03679\tHG03680\tHG03681\tHG03684\tHG03685\tHG03686\tHG03687\tHG03689\tHG03690\tHG03691\tHG03692\tHG03693\tHG03694\tHG03695\tHG03696\tHG03697\tHG03698\tHG03702\tHG03703\tHG03705\tHG03706\tHG03708\tHG03709\tHG03711\tHG03713\tHG03714\tHG03716\tHG03717\tHG03718\tHG03720\tHG03722\tHG03727\tHG03729\tHG03730\tHG03731\tHG03733\tHG03736\tHG03738\tHG03740\tHG03741\tHG03742\tHG03743\tHG03744\tHG03745\tHG03746\tHG03750\tHG03752\tHG03753\tHG03754\tHG03755\tHG03756\tHG03757\tHG03760\tHG03762\tHG03765\tHG03767\tHG03770\tHG03771\tHG03772\tHG03773\tHG03774\tHG03775\tHG03777\tHG03778\tHG03779\tHG03780\tHG03781\tHG03782\tHG03784\tHG03785\tHG03786\tHG03787\tHG03788\tHG03789\tHG03790\tHG03792\tHG03793\tHG03796\tHG03800\tHG03802\tHG03803\tHG03805\tHG03808\tHG03809\tHG03812\tHG03814\tHG03815\tHG03817\tHG03821\tHG03823\tHG03824\tHG03826\tHG03829\tHG03830\tHG03832\tHG03833\tHG03836\tHG03837\tHG03838\tHG03844\tHG03846\tHG03848\tHG03849\tHG03850\tHG03851\tHG03854\tHG03856\tHG03857\tHG03858\tHG03861\tHG03862\tHG03863\tHG03864\tHG03866\tHG03867\tHG03868\tHG03869\tHG03870\tHG03871\tHG03872\tHG03873\tHG03874\tHG03875\tHG03882\tHG03884\tHG03885\tHG03886\tHG03887\tHG03888\tHG03890\tHG03894\tHG03895\tHG03896\tHG03897\tHG03898\tHG03899\tHG03900\tHG03902\tHG03905\tHG03907\tHG03908\tHG03910\tHG03911\tHG03913\tHG03914\tHG03916\tHG03917\tHG03919\tHG03920\tHG03922\tHG03925\tHG03926\tHG03928\tHG03931\tHG03934\tHG03937\tHG03940\tHG03941\tHG03943\tHG03944\tHG03945\tHG03947\tHG03949\tHG03950\tHG03951\tHG03953\tHG03955\tHG03960\tHG03963\tHG03965\tHG03967\tHG03968\tHG03969\tHG03971\tHG03973\tHG03974\tHG03976\tHG03977\tHG03978\tHG03985\tHG03986\tHG03989\tHG03990\tHG03991\tHG03995\tHG03998\tHG03999\tHG04001\tHG04002\tHG04003\tHG04006\tHG04014\tHG04015\tHG04017\tHG04018\tHG04019\tHG04020\tHG04022\tHG04023\tHG04025\tHG04026\tHG04029\tHG04033\tHG04035\tHG04038\tHG04039\tHG04042\tHG04047\tHG04054\tHG04056\tHG04059\tHG04060\tHG04061\tHG04062\tHG04063\tHG04070\tHG04075\tHG04076\tHG04080\tHG04090\tHG04093\tHG04094\tHG04096\tHG04098\tHG04099\tHG04100\tHG04106\tHG04107\tHG04118\tHG04131\tHG04134\tHG04140\tHG04141\tHG04144\tHG04146\tHG04152\tHG04153\tHG04155\tHG04156\tHG04158\tHG04159\tHG04161\tHG04162\tHG04164\tHG04171\tHG04173\tHG04176\tHG04177\tHG04180\tHG04182\tHG04183\tHG04185\tHG04186\tHG04188\tHG04189\tHG04194\tHG04195\tHG04198\tHG04200\tHG04202\tHG04206\tHG04209\tHG04210\tHG04211\tHG04212\tHG04214\tHG04216\tHG04219\tHG04222\tHG04225\tHG04227\tHG04229\tHG04235\tHG04238\tHG04239\tHG04301\tHG04302\tHG04303\tNA06984\tNA06985\tNA06986\tNA06989\tNA06994\tNA07000\tNA07037\tNA07048\tNA07051\tNA07056\tNA07347\tNA07357\tNA10847\tNA10851\tNA11829\tNA11830\tNA11831\tNA11832\tNA11840\tNA11843\tNA11881\tNA11892\tNA11893\tNA11894\tNA11918\tNA11919\tNA11920\tNA11930\tNA11931\tNA11932\tNA11933\tNA11992\tNA11994\tNA11995\tNA12003\tNA12004\tNA12005\tNA12006\tNA12043\tNA12044\tNA12045\tNA12046\tNA12058\tNA12144\tNA12154\tNA12155\tNA12156\tNA12234\tNA12249\tNA12272\tNA12273\tNA12275\tNA12282\tNA12283\tNA12286\tNA12287\tNA12340\tNA12341\tNA12342\tNA12347\tNA12348\tNA12383\tNA12399\tNA12400\tNA12413\tNA12414\tNA12489\tNA12546\tNA12716\tNA12717\tNA12718\tNA12748\tNA12749\tNA12750\tNA12751\tNA12760\tNA12761\tNA12762\tNA12763\tNA12775\tNA12776\tNA12777\tNA12778\tNA12812\tNA12813\tNA12814\tNA12815\tNA12827\tNA12828\tNA12829\tNA12830\tNA12842\tNA12843\tNA12872\tNA12873\tNA12874\tNA12878\tNA12889\tNA12890\tNA18486\tNA18488\tNA18489\tNA18499\tNA18501\tNA18502\tNA18504\tNA18505\tNA18507\tNA18508\tNA18510\tNA18511\tNA18516\tNA18517\tNA18519\tNA18520\tNA18522\tNA18523\tNA18525\tNA18526\tNA18527\tNA18528\tNA18530\tNA18531\tNA18532\tNA18533\tNA18534\tNA18535\tNA18536\tNA18537\tNA18538\tNA18539\tNA18541\tNA18542\tNA18543\tNA18544\tNA18545\tNA18546\tNA18547\tNA18548\tNA18549\tNA18550\tNA18552\tNA18553\tNA18555\tNA18557\tNA18558\tNA18559\tNA18560\tNA18561\tNA18562\tNA18563\tNA18564\tNA18565\tNA18566\tNA18567\tNA18570\tNA18571\tNA18572\tNA18573\tNA18574\tNA18576\tNA18577\tNA18579\tNA18582\tNA18591\tNA18592\tNA18593\tNA18595\tNA18596\tNA18597\tNA18599\tNA18602\tNA18603\tNA18605\tNA18606\tNA18608\tNA18609\tNA18610\tNA18611\tNA18612\tNA18613\tNA18614\tNA18615\tNA18616\tNA18617\tNA18618\tNA18619\tNA18620\tNA18621\tNA18622\tNA18623\tNA18624\tNA18625\tNA18626\tNA18627\tNA18628\tNA18629\tNA18630\tNA18631\tNA18632\tNA18633\tNA18634\tNA18635\tNA18636\tNA18637\tNA18638\tNA18639\tNA18640\tNA18641\tNA18642\tNA18643\tNA18644\tNA18645\tNA18646\tNA18647\tNA18648\tNA18740\tNA18745\tNA18747\tNA18748\tNA18749\tNA18757\tNA18791\tNA18853\tNA18856\tNA18858\tNA18861\tNA18864\tNA18865\tNA18867\tNA18868\tNA18870\tNA18871\tNA18873\tNA18874\tNA18876\tNA18877\tNA18878\tNA18879\tNA18881\tNA18907\tNA18908\tNA18909\tNA18910\tNA18912\tNA18915\tNA18916\tNA18917\tNA18923\tNA18924\tNA18933\tNA18934\tNA18939\tNA18940\tNA18941\tNA18942\tNA18943\tNA18944\tNA18945\tNA18946\tNA18947\tNA18948\tNA18949\tNA18950\tNA18951\tNA18952\tNA18953\tNA18954\tNA18955\tNA18956\tNA18957\tNA18959\tNA18960\tNA18961\tNA18962\tNA18963\tNA18964\tNA18965\tNA18966\tNA18967\tNA18968\tNA18969\tNA18970\tNA18971\tNA18972\tNA18973\tNA18974\tNA18975\tNA18976\tNA18977\tNA18978\tNA18979\tNA18980\tNA18981\tNA18982\tNA18983\tNA18984\tNA18985\tNA18986\tNA18987\tNA18988\tNA18989\tNA18990\tNA18991\tNA18992\tNA18993\tNA18994\tNA18995\tNA18997\tNA18998\tNA18999\tNA19000\tNA19001\tNA19002\tNA19003\tNA19004\tNA19005\tNA19006\tNA19007\tNA19009\tNA19010\tNA19011\tNA19012\tNA19017\tNA19019\tNA19020\tNA19023\tNA19024\tNA19025\tNA19026\tNA19027\tNA19028\tNA19030\tNA19031\tNA19035\tNA19036\tNA19037\tNA19038\tNA19041\tNA19042\tNA19043\tNA19044\tNA19054\tNA19055\tNA19056\tNA19057\tNA19058\tNA19059\tNA19060\tNA19062\tNA19063\tNA19064\tNA19065\tNA19066\tNA19067\tNA19068\tNA19070\tNA19072\tNA19074\tNA19075\tNA19076\tNA19077\tNA19078\tNA19079\tNA19080\tNA19081\tNA19082\tNA19083\tNA19084\tNA19085\tNA19086\tNA19087\tNA19088\tNA19089\tNA19090\tNA19091\tNA19092\tNA19093\tNA19095\tNA19096\tNA19098\tNA19099\tNA19102\tNA19107\tNA19108\tNA19113\tNA19114\tNA19116\tNA19117\tNA19118\tNA19119\tNA19121\tNA19129\tNA19130\tNA19131\tNA19137\tNA19138\tNA19141\tNA19143\tNA19144\tNA19146\tNA19147\tNA19149\tNA19152\tNA19153\tNA19159\tNA19160\tNA19171\tNA19172\tNA19175\tNA19184\tNA19185\tNA19189\tNA19190\tNA19197\tNA19198\tNA19200\tNA19201\tNA19204\tNA19206\tNA19207\tNA19209\tNA19210\tNA19213\tNA19214\tNA19222\tNA19223\tNA19225\tNA19235\tNA19236\tNA19238\tNA19239\tNA19247\tNA19248\tNA19256\tNA19257\tNA19307\tNA19308\tNA19309\tNA19310\tNA19312\tNA19314\tNA19315\tNA19316\tNA19317\tNA19318\tNA19319\tNA19320\tNA19321\tNA19323\tNA19324\tNA19327\tNA19328\tNA19331\tNA19332\tNA19334\tNA19338\tNA19346\tNA19347\tNA19350\tNA19351\tNA19355\tNA19359\tNA19360\tNA19371\tNA19372\tNA19374\tNA19375\tNA19376\tNA19377\tNA19378\tNA19379\tNA19380\tNA19383\tNA19384\tNA19385\tNA19390\tNA19391\tNA19393\tNA19394\tNA19395\tNA19397\tNA19398\tNA19399\tNA19401\tNA19403\tNA19404\tNA19428\tNA19429\tNA19430\tNA19431\tNA19434\tNA19435\tNA19436\tNA19437\tNA19438\tNA19439\tNA19440\tNA19443\tNA19445\tNA19446\tNA19448\tNA19449\tNA19451\tNA19452\tNA19454\tNA19455\tNA19456\tNA19457\tNA19461\tNA19462\tNA19463\tNA19466\tNA19467\tNA19468\tNA19471\tNA19472\tNA19473\tNA19474\tNA19475\tNA19625\tNA19648\tNA19649\tNA19651\tNA19652\tNA19654\tNA19655\tNA19657\tNA19658\tNA19661\tNA19663\tNA19664\tNA19669\tNA19670\tNA19676\tNA19678\tNA19679\tNA19681\tNA19682\tNA19684\tNA19700\tNA19701\tNA19703\tNA19704\tNA19707\tNA19711\tNA19712\tNA19713\tNA19716\tNA19717\tNA19719\tNA19720\tNA19722\tNA19723\tNA19725\tNA19726\tNA19728\tNA19729\tNA19731\tNA19732\tNA19734\tNA19735\tNA19740\tNA19741\tNA19746\tNA19747\tNA19749\tNA19750\tNA19752\tNA19755\tNA19756\tNA19758\tNA19759\tNA19761\tNA19762\tNA19764\tNA19770\tNA19771\tNA19773\tNA19774\tNA19776\tNA19777\tNA19779\tNA19780\tNA19782\tNA19783\tNA19785\tNA19786\tNA19788\tNA19789\tNA19792\tNA19794\tNA19795\tNA19818\tNA19819\tNA19834\tNA19835\tNA19900\tNA19901\tNA19904\tNA19908\tNA19909\tNA19913\tNA19914\tNA19916\tNA19917\tNA19920\tNA19921\tNA19922\tNA19923\tNA19982\tNA19984\tNA20126\tNA20127\tNA20274\tNA20276\tNA20278\tNA20281\tNA20282\tNA20287\tNA20289\tNA20291\tNA20294\tNA20296\tNA20298\tNA20299\tNA20314\tNA20317\tNA20318\tNA20320\tNA20321\tNA20332\tNA20334\tNA20339\tNA20340\tNA20342\tNA20346\tNA20348\tNA20351\tNA20355\tNA20356\tNA20357\tNA20359\tNA20362\tNA20412\tNA20502\tNA20503\tNA20504\tNA20505\tNA20506\tNA20507\tNA20508\tNA20509\tNA20510\tNA20511\tNA20512\tNA20513\tNA20514\tNA20515\tNA20516\tNA20517\tNA20518\tNA20519\tNA20520\tNA20521\tNA20522\tNA20524\tNA20525\tNA20527\tNA20528\tNA20529\tNA20530\tNA20531\tNA20532\tNA20533\tNA20534\tNA20535\tNA20536\tNA20537\tNA20538\tNA20539\tNA20540\tNA20541\tNA20542\tNA20543\tNA20544\tNA20581\tNA20582\tNA20585\tNA20586\tNA20587\tNA20588\tNA20589\tNA20752\tNA20753\tNA20754\tNA20755\tNA20756\tNA20757\tNA20758\tNA20759\tNA20760\tNA20761\tNA20762\tNA20763\tNA20764\tNA20765\tNA20766\tNA20767\tNA20768\tNA20769\tNA20770\tNA20771\tNA20772\tNA20773\tNA20774\tNA20775\tNA20778\tNA20783\tNA20785\tNA20786\tNA20787\tNA20790\tNA20792\tNA20795\tNA20796\tNA20797\tNA20798\tNA20799\tNA20800\tNA20801\tNA20802\tNA20803\tNA20804\tNA20805\tNA20806\tNA20807\tNA20808\tNA20809\tNA20810\tNA20811\tNA20812\tNA20813\tNA20814\tNA20815\tNA20816\tNA20818\tNA20819\tNA20821\tNA20822\tNA20826\tNA20827\tNA20828\tNA20829\tNA20831\tNA20832\tNA20845\tNA20846\tNA20847\tNA20849\tNA20850\tNA20851\tNA20852\tNA20853\tNA20854\tNA20856\tNA20858\tNA20859\tNA20861\tNA20862\tNA20863\tNA20864\tNA20866\tNA20867\tNA20868\tNA20869\tNA20870\tNA20872\tNA20873\tNA20874\tNA20875\tNA20876\tNA20877\tNA20878\tNA20881\tNA20882\tNA20883\tNA20884\tNA20885\tNA20886\tNA20887\tNA20888\tNA20889\tNA20890\tNA20891\tNA20892\tNA20894\tNA20895\tNA20896\tNA20897\tNA20899\tNA20900\tNA20901\tNA20902\tNA20903\tNA20904\tNA20905\tNA20906\tNA20908\tNA20910\tNA20911\tNA21086\tNA21087\tNA21088\tNA21089\tNA21090\tNA21091\tNA21092\tNA21093\tNA21094\tNA21095\tNA21097\tNA21098\tNA21099\tNA21100\tNA21101\tNA21102\tNA21103\tNA21104\tNA21105\tNA21106\tNA21107\tNA21108\tNA21109\tNA21110\tNA21111\tNA21112\tNA21113\tNA21114\tNA21115\tNA21116\tNA21117\tNA21118\tNA21119\tNA21120\tNA21121\tNA21122\tNA21123\tNA21124\tNA21125\tNA21126\tNA21127\tNA21128\tNA21129\tNA21130\tNA21133\tNA21135\tNA21137\tNA21141\tNA21142\tNA21143\tNA21144\n",
            "22\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Command: bcftools view [OPTIONS] file.vcf.gz [REGION […​]]\n",
        "<br>\n",
        "<br>\n",
        "Usage: View, subset and filter VCF or BCF files by position and filtering expression. Convert between VCF and BCF. Former bcftools subset.\n",
        "\n",
        "------------------------------------\n",
        "\n",
        "The -h tag makes the output of the view command only include headers.\n",
        "<br>\n",
        "So this is the last line of the output:\n",
        "<br>\n",
        "CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tHG...\n",
        "NA...\n",
        "\n",
        "\n",
        "------------------------------------\n",
        "Column descriptions:\n",
        "<br>\n",
        "CHROM: Number of chromosome of the SNPs\n",
        "<br>\n",
        "POS: position (index) of SNP on\n",
        "chromosome\n",
        "<br>\n",
        "ID: SNP ID (e.g. could be found in SNPedia or DBSNP)\n",
        "<br>\n",
        "REF: Reference (more common)‌allele\n",
        "<br>\n",
        "ALT: Alternative alleles\n",
        "<br>\n",
        "QUAL: Quality score of SNP (measuring confidence)\n",
        "<br>\n",
        "INFO: additional information\n",
        "<br>\n",
        "FORMAT: Genotype formats\n",
        "<br>\n",
        "HG... or NA...\n",
        "Each HG.. or NA... column represents one human sample's Genotype info.\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "5lTDnLlRY4SD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Part 2\n",
        "!bcftools view -m2 -M2 -v snps {unphased_data} -o {folder}snp2allele.vcf.gz\n",
        "\n",
        "# Part 3\n",
        "!bcftools view -i 'MAF[0] > 0.01' -o {folder}snp_maf.vcf.gz {folder}snp2allele.vcf.gz\n",
        "\n",
        "# Part 4 - step 1\n",
        "# Choose first 100 snps\n",
        "!bcftools view {folder}snp_maf.vcf.gz | head -n 122 > snps1to100withheader.vcf"
      ],
      "metadata": {
        "id": "pK5kAMEaZPby"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Part 2, 3 and 4A\n",
        "\n",
        "---------------------------\n",
        "\n",
        "<br>\n",
        "The 1st command:\n",
        "<br>\n",
        "bcftools options:\n",
        "<br>\n",
        "-m2 -M2: filters variants with only 2 alleles\n",
        "<br>\n",
        "-v snps: filter only SNPs (not multiple nucleotide variations)\n",
        "\n",
        "---------------------------\n",
        "\n",
        "<br>\n",
        "The 2nd command:\n",
        "<br>\n",
        "bcftools options:\n",
        "<br>\n",
        "'MAF[0] > 0.01': filters snps with Minor allele frequency (MAF) > 0.01 (1%)\n",
        "(in other wordss, common variants)\n",
        "<br>\n",
        "\n",
        "\n",
        "---------------------------\n",
        "\n",
        "<br>\n",
        "The 3rd command:\n",
        "<br>\n",
        "bcftools options:\n",
        "<br>\n",
        "head -n: writes the first 122 lines to output.\n",
        "<br>\n",
        "Why 122?\n",
        "because the 'wc -l' command output had a '22' in the end, which means the header takes the first 22 lines, hence 22 header rows + 100 snp rows.\n"
      ],
      "metadata": {
        "id": "ykZGzfLHe5Sm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Part 4 - step 2\n",
        "# download, extract & run shapeit exe\n",
        "!wget https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.v2.r904.glibcv2.17.linux.tar.gz\n",
        "!tar -zxvf /content/shapeit.v2.r904.glibcv2.17.linux.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "12GuEXbiXOSB",
        "outputId": "4d98077b-2b21-44a6-c532-16b04ec42d18"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-12-28 22:27:49--  https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.v2.r904.glibcv2.17.linux.tar.gz\n",
            "Resolving mathgen.stats.ox.ac.uk (mathgen.stats.ox.ac.uk)... 163.1.210.102\n",
            "Connecting to mathgen.stats.ox.ac.uk (mathgen.stats.ox.ac.uk)|163.1.210.102|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 13047276 (12M) [application/x-gzip]\n",
            "Saving to: ‘shapeit.v2.r904.glibcv2.17.linux.tar.gz’\n",
            "\n",
            "shapeit.v2.r904.gli 100%[===================>]  12.44M  5.64MB/s    in 2.2s    \n",
            "\n",
            "2024-12-28 22:27:52 (5.64 MB/s) - ‘shapeit.v2.r904.glibcv2.17.linux.tar.gz’ saved [13047276/13047276]\n",
            "\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/bin/\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/bin/shapeit\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/LICENCE\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/bamlist\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/NA12891.bam\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/NA12891.bam.bai\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/NA12892.bam\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/NA12892.bam.bai\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/genetic_map.chrX.txt\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/genetic_map.txt\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.bed\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.bim\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.chrX.bed\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.chrX.bim\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.chrX.fam\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.fam\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.gen.gz\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.map\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.ped\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.sample\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.sets\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.subset.inds\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.subset.site\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/gwas.vcf.gz\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/reference.chrX.haplotypes.gz\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/reference.chrX.legend.gz\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/reference.chrX.sample\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/reference.haplotypes.gz\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/reference.legend.gz\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/reference.sample\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/sequence.multiple.vcf.gz\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/sequence.single.inds.pir\n",
            "shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/sequence.single.vcf.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        " # Part 4 - step 3\n",
        "shapeit_v2 = '/content/shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/bin/shapeit'\n",
        "\n",
        "# The actual Phasing\n",
        "!{shapeit_v2} --input-vcf /content/snps1to100withheader.vcf --output-max /content/phasing_output.vcf.gz\n",
        "\n",
        "# Same as above but the output format isn't .gz anymore\n",
        "!{shapeit_v2} -V /content/snps1to100withheader.vcf -O phasing_output.haps\n",
        "\n",
        "# Convert .haps to .vcf\n",
        "!{shapeit_v2} -convert --input-haps /content/phasing_output.haps --output-vcf /content/phasing_output.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "r18SRKnAkAva",
        "outputId": "4bd5a848-3687-4f00-d8c6-381fb9be13fa"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\u001b[33mS\u001b[0megmented \u001b[33mHAP\u001b[0mlotype \u001b[33mE\u001b[0mstimation & \u001b[33mI\u001b[0mmputation \u001b[33mT\u001b[0mool\n",
            "  * Authors : Olivier Delaneau, Jared O'Connell, Jean-François Zagury, Jonathan Marchini\n",
            "  * Contact : send an email to the OXSTATGEN mail list https://www.jiscmail.ac.uk/cgi-bin/webadmin?A0=OXSTATGEN\n",
            "  * Webpage : https://mathgen.stats.ox.ac.uk/shapeit\n",
            "  * Version : v2.r904\n",
            "  * Date    : 29/12/2024 00:50:25\n",
            "  * LOGfile : [shapeit_29122024_00h50m25s_126896c3-692a-4039-9247-3ade66738ae4.log]\n",
            "\n",
            "MODE \u001b[33m-phase\u001b[0m : PHASING GENOTYPE DATA\n",
            "  * Autosome (chr1 ... chr22)\n",
            "  * Window-based model (SHAPEIT v2)\n",
            "  * MCMC iteration\n",
            "\n",
            "Parameters :\n",
            "  * Seed : 1735433425\n",
            "  * Parallelisation: 1 threads\n",
            "  * Ref allele is NOT aligned on the reference genome\n",
            "  * MCMC: 35 iterations [7 B + 1 runs of 8 P + 20 M]\n",
            "  * Model: 100 states per window [100 H + 0 PM + 0 R + 0 COV ] / Windows of ~2.0 Mb / rho = 0.00040\n",
            "\n",
            "Reading unphased data in [/content/snps1to100withheader.vcf]\n",
            "  * 2548 samples included\n",
            "  * 2548 unrelateds / 0 duos / 0 trios in 2548 different families\n",
            "  * 98 SNPs included\n",
            "\n",
            "Checking missingness and MAF...\n",
            "\n",
            "Building graphs [2548/2548]\n",
            "  * 2548 graphs / 11756 segments / ~21 SNPs per segment / 276997 transitions\n",
            "  * 0 haploids / 2548 unrelateds / 0 duos / 0 trios\n",
            "  * 5096 founder haplotypes\n",
            "\n",
            "Sampling haplotypes [2548/2548]\n",
            "\n",
            "Burn-in iteration [1/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [2/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [3/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [4/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [5/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [6/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [7/7] [2548/2548]\n",
            "\n",
            "Pruning iteration [1/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [2/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [3/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [4/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [5/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [6/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [7/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [8/8] [2548/2548]\n",
            "\n",
            "Pruning graphs [2548/2548]\n",
            "  * 2548 graphs / 6098 segments / ~40 SNPs per segment / 103580 transitions\n",
            "  * 0 haploids / 2548 unrelateds / 0 duos / 0 trios\n",
            "  * 5096 founder haplotypes\n",
            "\n",
            "Main iteration [1/20] [2548/2548]\n",
            "\n",
            "Main iteration [2/20] [2548/2548]\n",
            "\n",
            "Main iteration [3/20] [2548/2548]\n",
            "\n",
            "Main iteration [4/20] [2548/2548]\n",
            "\n",
            "Main iteration [5/20] [2548/2548]\n",
            "\n",
            "Main iteration [6/20] [2548/2548]\n",
            "\n",
            "Main iteration [7/20] [2548/2548]\n",
            "\n",
            "Main iteration [8/20] [2548/2548]\n",
            "\n",
            "Main iteration [9/20] [2548/2548]\n",
            "\n",
            "Main iteration [10/20] [2548/2548]\n",
            "\n",
            "Main iteration [11/20] [2548/2548]\n",
            "\n",
            "Main iteration [12/20] [2548/2548]\n",
            "\n",
            "Main iteration [13/20] [2548/2548]\n",
            "\n",
            "Main iteration [14/20] [2548/2548]\n",
            "\n",
            "Main iteration [15/20] [2548/2548]\n",
            "\n",
            "Main iteration [16/20] [2548/2548]\n",
            "\n",
            "Main iteration [17/20] [2548/2548]\n",
            "\n",
            "Main iteration [18/20] [2548/2548]\n",
            "\n",
            "Main iteration [19/20] [2548/2548]\n",
            "\n",
            "Main iteration [20/20] [2548/2548]\n",
            "\n",
            "Normalising graphs [2548/2548]\n",
            "\n",
            "Solving haplotypes [2548/2548]\n",
            "\n",
            "Writing sample list in [/content/phasing_output.vcf.gz.sample]\n",
            "\n",
            "Writing site list and haplotypes in [/content/phasing_output.vcf.gz.haps]\n",
            "\n",
            "Running time: 125 seconds\n",
            "\n",
            "\u001b[33mS\u001b[0megmented \u001b[33mHAP\u001b[0mlotype \u001b[33mE\u001b[0mstimation & \u001b[33mI\u001b[0mmputation \u001b[33mT\u001b[0mool\n",
            "  * Authors : Olivier Delaneau, Jared O'Connell, Jean-François Zagury, Jonathan Marchini\n",
            "  * Contact : send an email to the OXSTATGEN mail list https://www.jiscmail.ac.uk/cgi-bin/webadmin?A0=OXSTATGEN\n",
            "  * Webpage : https://mathgen.stats.ox.ac.uk/shapeit\n",
            "  * Version : v2.r904\n",
            "  * Date    : 29/12/2024 00:52:30\n",
            "  * LOGfile : [shapeit_29122024_00h52m30s_42cfcaf4-b308-4151-b351-96d152d54e7f.log]\n",
            "\n",
            "MODE \u001b[33m-phase\u001b[0m : PHASING GENOTYPE DATA\n",
            "  * Autosome (chr1 ... chr22)\n",
            "  * Window-based model (SHAPEIT v2)\n",
            "  * MCMC iteration\n",
            "\n",
            "Parameters :\n",
            "  * Seed : 1735433550\n",
            "  * Parallelisation: 1 threads\n",
            "  * Ref allele is NOT aligned on the reference genome\n",
            "  * MCMC: 35 iterations [7 B + 1 runs of 8 P + 20 M]\n",
            "  * Model: 100 states per window [100 H + 0 PM + 0 R + 0 COV ] / Windows of ~2.0 Mb / rho = 0.00040\n",
            "\n",
            "Reading unphased data in [/content/snps1to100withheader.vcf]\n",
            "  * 2548 samples included\n",
            "  * 2548 unrelateds / 0 duos / 0 trios in 2548 different families\n",
            "  * 98 SNPs included\n",
            "\n",
            "Checking missingness and MAF...\n",
            "\n",
            "Building graphs [2548/2548]\n",
            "  * 2548 graphs / 11756 segments / ~21 SNPs per segment / 276997 transitions\n",
            "  * 0 haploids / 2548 unrelateds / 0 duos / 0 trios\n",
            "  * 5096 founder haplotypes\n",
            "\n",
            "Sampling haplotypes [2548/2548]\n",
            "\n",
            "Burn-in iteration [1/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [2/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [3/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [4/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [5/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [6/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [7/7] [2548/2548]\n",
            "\n",
            "Pruning iteration [1/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [2/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [3/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [4/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [5/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [6/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [7/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [8/8] [2548/2548]\n",
            "\n",
            "Pruning graphs [2548/2548]\n",
            "  * 2548 graphs / 6292 segments / ~39 SNPs per segment / 107581 transitions\n",
            "  * 0 haploids / 2548 unrelateds / 0 duos / 0 trios\n",
            "  * 5096 founder haplotypes\n",
            "\n",
            "Main iteration [1/20] [2548/2548]\n",
            "\n",
            "Main iteration [2/20] [2548/2548]\n",
            "\n",
            "Main iteration [3/20] [2548/2548]\n",
            "\n",
            "Main iteration [4/20] [2548/2548]\n",
            "\n",
            "Main iteration [5/20] [2548/2548]\n",
            "\n",
            "Main iteration [6/20] [2548/2548]\n",
            "\n",
            "Main iteration [7/20] [2548/2548]\n",
            "\n",
            "Main iteration [8/20] [2548/2548]\n",
            "\n",
            "Main iteration [9/20] [2548/2548]\n",
            "\n",
            "Main iteration [10/20] [2548/2548]\n",
            "\n",
            "Main iteration [11/20] [2548/2548]\n",
            "\n",
            "Main iteration [12/20] [2548/2548]\n",
            "\n",
            "Main iteration [13/20] [2548/2548]\n",
            "\n",
            "Main iteration [14/20] [2548/2548]\n",
            "\n",
            "Main iteration [15/20] [2548/2548]\n",
            "\n",
            "Main iteration [16/20] [2548/2548]\n",
            "\n",
            "Main iteration [17/20] [2548/2548]\n",
            "\n",
            "Main iteration [18/20] [2548/2548]\n",
            "\n",
            "Main iteration [19/20] [2548/2548]\n",
            "\n",
            "Main iteration [20/20] [2548/2548]\n",
            "\n",
            "Normalising graphs [2548/2548]\n",
            "\n",
            "Solving haplotypes [2548/2548]\n",
            "\n",
            "Writing sample list in [phasing_output.haps.sample]\n",
            "\n",
            "Writing site list and haplotypes in [phasing_output.haps.haps]\n",
            "\n",
            "Running time: 131 seconds\n",
            "\n",
            "\u001b[33mS\u001b[0megmented \u001b[33mHAP\u001b[0mlotype \u001b[33mE\u001b[0mstimation & \u001b[33mI\u001b[0mmputation \u001b[33mT\u001b[0mool\n",
            "  * Authors : Olivier Delaneau, Jared O'Connell, Jean-François Zagury, Jonathan Marchini\n",
            "  * Contact : send an email to the OXSTATGEN mail list https://www.jiscmail.ac.uk/cgi-bin/webadmin?A0=OXSTATGEN\n",
            "  * Webpage : https://mathgen.stats.ox.ac.uk/shapeit\n",
            "  * Version : v2.r904\n",
            "  * Date    : 29/12/2024 00:54:41\n",
            "  * LOGfile : [shapeit_29122024_00h54m41s_16b80905-8910-4c28-9690-4729a785e568.log]\n",
            "\n",
            "MODE \u001b[33m-convert\u001b[0m : CONVERTING/SUBSETTING SHAPEIT HAPLOTYPES\n",
            "\n",
            "Parameters :\n",
            "  * Seed : 1735433681\n",
            "  * Parallelisation: 1 threads\n",
            "  * Ref allele is NOT aligned on the reference genome\n",
            "\n",
            "Reading sample list in [/content/phasing_output.haps.sample]\n",
            "  * 2548 individuals included\n",
            "\n",
            "Reading site list and haplotypes in [/content/phasing_output.haps.haps]\n",
            "  * 98 sites included\n",
            "\n",
            "Mapping sites and samples of the haplotype set\n",
            "  * 98 sites found\n",
            "  * 0 sites not found\n",
            "  * 5096 founder haplotypes\n",
            "\n",
            "Writing haplotypes in [/content/phasing_output.vcf] in VCF format\n",
            "\n",
            "Running time: 0 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Part 5\n",
        "# Compare with ground truth for one sample\n",
        "\n",
        "# get ground truth data\n",
        "!bcftools view -m2 -M2 -v snps '/content/DataHW03/chr14..GRCh38.phased.vcf.gz' -o /content/all_labels.vcf.gz\n",
        "!bcftools view all_labels.vcf.gz | head -n 122 > labels.vcf"
      ],
      "metadata": {
        "id": "wxRs2oK304A2"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Part 5- Cont'd\n",
        "\n",
        "# read ground truth data\n",
        "!bcftools query -s HG00097 -f '[%CHROM\\t%POS\\t%REF\\t%ALT\\t%GT\\n]' /content/labels.vcf\n",
        "\n",
        "# compare\n",
        "!whatshap compare --names truth, whatshap --tsv-pairwise eval.tsv --sample=HG00097 /content/labels.vcf /content/phasing_output.vcf\n",
        "\n",
        "!whatshap compare --sample=HG00097 /content/labels.vcf /content/phasing_output.vcf\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "n3aeaATy_RGL",
        "outputId": "d83312e1-be7c-4d52-8897-927dfed1979a"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "14\t16038675\tC\tT\t0|0\n",
            "14\t16055980\tC\tT\t0|0\n",
            "14\t16055981\tG\tA\t0|0\n",
            "14\t16055998\tC\tT\t0|0\n",
            "14\t16056028\tC\tT\t0|0\n",
            "14\t16056029\tG\tA\t0|0\n",
            "14\t16056043\tG\tC\t0|0\n",
            "14\t16056069\tC\tT\t0|0\n",
            "14\t16056115\tC\tT\t0|0\n",
            "14\t16056704\tG\tA\t0|0\n",
            "14\t16058840\tC\tT\t0|0\n",
            "14\t16076314\tT\tC\t0|0\n",
            "14\t16076317\tA\tG\t0|0\n",
            "14\t16091887\tA\tG\t0|0\n",
            "14\t16091892\tG\tT\t0|0\n",
            "14\t16091894\tG\tA\t0|0\n",
            "14\t16091900\tT\tG\t0|0\n",
            "14\t16091915\tA\tC\t0|0\n",
            "14\t16091918\tC\tA\t0|0\n",
            "14\t16091937\tG\tT\t0|0\n",
            "14\t16091948\tT\tC\t0|0\n",
            "14\t16092003\tC\tT\t0|0\n",
            "14\t16092027\tC\tT\t0|0\n",
            "14\t16092120\tA\tG\t0|0\n",
            "14\t16092139\tT\tA\t0|0\n",
            "14\t16092163\tT\tC\t0|0\n",
            "14\t16092171\tT\tG\t0|0\n",
            "14\t16092174\tA\tT\t0|0\n",
            "14\t16092228\tT\tC\t0|0\n",
            "14\t16092274\tG\tA\t0|0\n",
            "14\t16092294\tC\tA\t0|0\n",
            "14\t16092313\tT\tC\t0|0\n",
            "14\t16092318\tT\tG\t0|0\n",
            "14\t16092319\tG\tC\t0|0\n",
            "14\t16092325\tT\tG\t0|0\n",
            "14\t16092350\tA\tG\t0|0\n",
            "14\t16092352\tG\tC\t0|0\n",
            "14\t16092367\tC\tT\t0|0\n",
            "14\t16092368\tG\tA\t0|0\n",
            "14\t16092369\tA\tG\t0|0\n",
            "14\t16092378\tG\tC\t0|0\n",
            "14\t16092382\tA\tT\t0|0\n",
            "14\t16092384\tC\tT\t0|0\n",
            "14\t16092397\tT\tA\t0|0\n",
            "14\t16092426\tC\tT\t0|0\n",
            "14\t16092444\tT\tA\t0|0\n",
            "14\t16092448\tC\tT\t0|0\n",
            "14\t16092449\tG\tA\t0|0\n",
            "14\t16092465\tC\tT\t0|0\n",
            "14\t16092471\tA\tG\t0|0\n",
            "14\t16092477\tT\tA\t0|0\n",
            "14\t16092478\tG\tA\t0|0\n",
            "14\t16092480\tA\tG\t0|0\n",
            "14\t16092483\tT\tG\t0|0\n",
            "14\t16092498\tT\tC\t0|0\n",
            "14\t16092530\tT\tG\t0|0\n",
            "14\t16092562\tG\tA\t0|0\n",
            "14\t16092565\tT\tG\t0|0\n",
            "14\t16092568\tG\tT\t0|0\n",
            "14\t16092598\tT\tG\t0|0\n",
            "14\t16092604\tG\tA\t0|0\n",
            "14\t16092623\tA\tG\t0|0\n",
            "14\t16092652\tG\tT\t0|0\n",
            "14\t16092669\tG\tT\t0|0\n",
            "14\t16092708\tA\tT\t0|0\n",
            "14\t16092710\tG\tA\t0|0\n",
            "14\t16092711\tT\tC\t0|0\n",
            "14\t16092733\tT\tG\t0|0\n",
            "14\t16092736\tT\tC\t0|0\n",
            "14\t16092738\tT\tC\t0|0\n",
            "14\t16092769\tT\tG\t0|0\n",
            "14\t16092774\tT\tC\t0|0\n",
            "14\t16092777\tC\tG\t0|0\n",
            "14\t16092788\tG\tC\t0|0\n",
            "14\t16092797\tT\tA\t0|0\n",
            "14\t16092801\tG\tC\t0|0\n",
            "14\t16092876\tC\tT\t0|0\n",
            "14\t16092877\tG\tA\t0|0\n",
            "14\t16092886\tG\tC\t0|0\n",
            "14\t16092896\tT\tA\t0|0\n",
            "14\t16092908\tG\tA\t0|0\n",
            "14\t16092934\tG\tA\t0|0\n",
            "14\t16092953\tG\tT\t0|0\n",
            "14\t16092984\tC\tT\t0|0\n",
            "14\t16092990\tG\tA\t0|0\n",
            "14\t16092995\tT\tA\t0|0\n",
            "14\t16093005\tT\tC\t0|0\n",
            "14\t16093044\tT\tC\t0|0\n",
            "14\t16093060\tA\tG\t0|0\n",
            "14\t16093079\tT\tC\t0|0\n",
            "14\t16093083\tC\tT\t0|0\n",
            "14\t16093172\tC\tT\t0|0\n",
            "14\t16093175\tT\tC\t0|0\n",
            "14\t16093177\tT\tC\t0|0\n",
            "14\t16093203\tT\tA\t0|0\n",
            "14\t16093253\tC\tG\t1|1\n",
            "14\t16093266\tG\tA\t0|0\n",
            "14\t16093269\tT\tA\t0|0\n",
            "14\t16093278\tC\tA\t0|0\n",
            "usage: whatshap [-h] [--version] [--debug] {compare} ...\n",
            "\n",
            "positional arguments:\n",
            "  {compare}\n",
            "    compare   Compare two or more phased variant files\n",
            "\n",
            "options:\n",
            "  -h, --help  show this help message and exit\n",
            "  --version   show program's version number and exit\n",
            "  --debug     Print debug messages\n",
            "whatshap: error: unrecognized arguments: /content/labels.vcf /content/phasing_output.vcf\n",
            "Comparing phasings for sample HG00097\n",
            "Reading phasing from '/content/labels.vcf'\n",
            "Reading phasing from '/content/phasing_output.vcf'\n",
            "[W::vcf_parse] Contig '14' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
            "Chromosomes present in all VCFs: 14\n",
            "FILENAMES\n",
            "  file0 = /content/labels.vcf\n",
            "  file1 = /content/phasing_output.vcf\n",
            "---------------- Chromosome 14 ----------------\n",
            "VARIANT COUNTS (heterozygous / all): \n",
            "              file0:         0 /        99\n",
            "              file1:         8 /        98\n",
            "              UNION:         8 /       174\n",
            "       INTERSECTION:         0 /        23\n",
            "PAIRWISE COMPARISON: file0 <--> file1:\n",
            "         common heterozygous variants:         0\n",
            "         (restricting to these below)\n",
            "        non-singleton blocks in file0:         0\n",
            "                 --> covered variants:         0\n",
            "        non-singleton blocks in file1:         0\n",
            "                 --> covered variants:         0\n",
            "    non-singleton intersection blocks:         0\n",
            "                 --> covered variants:         0\n",
            "              ALL INTERSECTION BLOCKS: ---------\n",
            "    phased pairs of variants assessed:         0\n",
            "                        switch errors:         0\n",
            "                    switch error rate:        --\n",
            "            switch/flip decomposition:       0/0\n",
            "                     switch/flip rate:        --\n",
            "          Block-wise Hamming distance:         0\n",
            "      Block-wise Hamming distance [%]:        --\n",
            "                  Different genotypes:         0\n",
            "              Different genotypes [%]:        --\n",
            "           LARGEST INTERSECTION BLOCK: ---------\n",
            "    phased pairs of variants assessed:         0\n",
            "                        switch errors:         0\n",
            "                    switch error rate:        --\n",
            "            switch/flip decomposition:       0/0\n",
            "                     switch/flip rate:        --\n",
            "                     Hamming distance:         0\n",
            "                 Hamming distance [%]:        --\n",
            "                  Different genotypes:         0\n",
            "              Different genotypes [%]:        --\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Part 6:\n",
        "\n",
        "-------------------------------------------\n",
        "heavily based on:\n",
        "https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#model\n",
        "<br>\n",
        "<br>\n",
        "\n",
        "States Conditioning:\n",
        "<br>\n",
        "You can increase accuracy of SHAPEIT by increasing the number of conditioning states on which haplotype estimation is based. (in HMMs, it is the length of the sequence of hidden states, e.g. the total number of days, whether sunny or rainy) By default, SHAPEIT uses 100 states per SNP across the dataset which gives good accuracy while maintaining reasonable running times. The complexity of the algorithm is linear with the number of conditioning states.\n",
        "<br>\n",
        "example:\n",
        "```\n",
        "shapeit -B gwas \\\n",
        "        -M genetic_map.txt \\\n",
        "        -O gwas.phased \\\n",
        "        --states 200\n",
        "```\n",
        "<br>\n",
        "<br>\n",
        "Window:\n",
        "<br>\n",
        "It is the length of the genotype region whose haplotypes we want to find by phasing. For GWAS dataset, we recommend a value of 2Mb, which should be fine for most application. By default, SHAPEIT2 uses windows of ~2Mb on average. Sometimes a value of 0.5Mb may give better results. To do this, use the --window option as follows:\n",
        "\n",
        "```\n",
        "shapeit -B gwas \\\n",
        "        -M genetic_map.txt \\\n",
        "        -O gwas.phased \\\n",
        "        --window 0.5\n",
        "```\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "YqT0Hk9UvpFR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Part 7 - like part 4 but using genetic map\n",
        "!{shapeit_v2} -V /content/snps1to100withheader.vcf -O /content/phasing_output_genetic_map.haps -M /content/shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/genetic_map.txt\n",
        "# convert .hap to .vcf\n",
        "!{shapeit_v2} -convert --input-haps /content/phasing_output_genetic_map.haps --output-vcf /content/phasing_output_genetic_map.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ELaRJl7hnx5f",
        "outputId": "2de3baec-8f81-4c3d-b154-be590ea7ae85"
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\u001b[33mS\u001b[0megmented \u001b[33mHAP\u001b[0mlotype \u001b[33mE\u001b[0mstimation & \u001b[33mI\u001b[0mmputation \u001b[33mT\u001b[0mool\n",
            "  * Authors : Olivier Delaneau, Jared O'Connell, Jean-François Zagury, Jonathan Marchini\n",
            "  * Contact : send an email to the OXSTATGEN mail list https://www.jiscmail.ac.uk/cgi-bin/webadmin?A0=OXSTATGEN\n",
            "  * Webpage : https://mathgen.stats.ox.ac.uk/shapeit\n",
            "  * Version : v2.r904\n",
            "  * Date    : 29/12/2024 01:16:21\n",
            "  * LOGfile : [shapeit_29122024_01h16m21s_33858148-96f7-4e68-a0a9-16f062de96ad.log]\n",
            "\n",
            "MODE \u001b[33m-phase\u001b[0m : PHASING GENOTYPE DATA\n",
            "  * Autosome (chr1 ... chr22)\n",
            "  * Window-based model (SHAPEIT v2)\n",
            "  * MCMC iteration\n",
            "\n",
            "Parameters :\n",
            "  * Seed : 1735434981\n",
            "  * Parallelisation: 1 threads\n",
            "  * Ref allele is NOT aligned on the reference genome\n",
            "  * MCMC: 35 iterations [7 B + 1 runs of 8 P + 20 M]\n",
            "  * Model: 100 states per window [100 H + 0 PM + 0 R + 0 COV ] / Windows of ~2.0 Mb / Ne = 15000\n",
            "\n",
            "Reading unphased data in [/content/snps1to100withheader.vcf]\n",
            "  * 2548 samples included\n",
            "  * 2548 unrelateds / 0 duos / 0 trios in 2548 different families\n",
            "  * 98 SNPs included\n",
            "\n",
            "Reading genetic map in [/content/shapeit.v2.904.3.10.0-693.11.6.el7.x86_64/example/genetic_map.txt]\n",
            "  * 79922 genetic positions found\n",
            "  * #set=0 / #interpolated=98\n",
            "  * Physical map [16.04 Mb -> 16.10 Mb] / Genetic map [39.05 cM -> 39.24 cM]\n",
            "\n",
            "Checking missingness and MAF...\n",
            "\n",
            "Building graphs [2548/2548]\n",
            "  * 2548 graphs / 11756 segments / ~21 SNPs per segment / 276997 transitions\n",
            "  * 0 haploids / 2548 unrelateds / 0 duos / 0 trios\n",
            "  * 5096 founder haplotypes\n",
            "\n",
            "Sampling haplotypes [2548/2548]\n",
            "\n",
            "Burn-in iteration [1/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [2/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [3/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [4/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [5/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [6/7] [2548/2548]\n",
            "\n",
            "Burn-in iteration [7/7] [2548/2548]\n",
            "\n",
            "Pruning iteration [1/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [2/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [3/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [4/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [5/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [6/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [7/8] [2548/2548]\n",
            "\n",
            "Pruning iteration [8/8] [2548/2548]\n",
            "\n",
            "Pruning graphs [2548/2548]\n",
            "  * 2548 graphs / 6061 segments / ~41 SNPs per segment / 100782 transitions\n",
            "  * 0 haploids / 2548 unrelateds / 0 duos / 0 trios\n",
            "  * 5096 founder haplotypes\n",
            "\n",
            "Main iteration [1/20] [2548/2548]\n",
            "\n",
            "Main iteration [2/20] [2548/2548]\n",
            "\n",
            "Main iteration [3/20] [2548/2548]\n",
            "\n",
            "Main iteration [4/20] [2548/2548]\n",
            "\n",
            "Main iteration [5/20] [2548/2548]\n",
            "\n",
            "Main iteration [6/20] [2548/2548]\n",
            "\n",
            "Main iteration [7/20] [2548/2548]\n",
            "\n",
            "Main iteration [8/20] [2548/2548]\n",
            "\n",
            "Main iteration [9/20] [2548/2548]\n",
            "\n",
            "Main iteration [10/20] [2548/2548]\n",
            "\n",
            "Main iteration [11/20] [2548/2548]\n",
            "\n",
            "Main iteration [12/20] [2548/2548]\n",
            "\n",
            "Main iteration [13/20] [2548/2548]\n",
            "\n",
            "Main iteration [14/20] [2548/2548]\n",
            "\n",
            "Main iteration [15/20] [2548/2548]\n",
            "\n",
            "Main iteration [16/20] [2548/2548]\n",
            "\n",
            "Main iteration [17/20] [2548/2548]\n",
            "\n",
            "Main iteration [18/20] [2548/2548]\n",
            "\n",
            "Main iteration [19/20] [2548/2548]\n",
            "\n",
            "Main iteration [20/20] [2548/2548]\n",
            "\n",
            "Normalising graphs [2548/2548]\n",
            "\n",
            "Solving haplotypes [2548/2548]\n",
            "\n",
            "Writing sample list in [/content/phasing_output_genetic_map.haps.sample]\n",
            "\n",
            "Writing site list and haplotypes in [/content/phasing_output_genetic_map.haps.haps]\n",
            "\n",
            "Running time: 131 seconds\n",
            "\n",
            "\u001b[33mS\u001b[0megmented \u001b[33mHAP\u001b[0mlotype \u001b[33mE\u001b[0mstimation & \u001b[33mI\u001b[0mmputation \u001b[33mT\u001b[0mool\n",
            "  * Authors : Olivier Delaneau, Jared O'Connell, Jean-François Zagury, Jonathan Marchini\n",
            "  * Contact : send an email to the OXSTATGEN mail list https://www.jiscmail.ac.uk/cgi-bin/webadmin?A0=OXSTATGEN\n",
            "  * Webpage : https://mathgen.stats.ox.ac.uk/shapeit\n",
            "  * Version : v2.r904\n",
            "  * Date    : 29/12/2024 01:18:32\n",
            "  * LOGfile : [shapeit_29122024_01h18m32s_63359e11-aae4-4186-9dee-855e5c692bdb.log]\n",
            "\n",
            "MODE \u001b[33m-convert\u001b[0m : CONVERTING/SUBSETTING SHAPEIT HAPLOTYPES\n",
            "\n",
            "Parameters :\n",
            "  * Seed : 1735435112\n",
            "  * Parallelisation: 1 threads\n",
            "  * Ref allele is NOT aligned on the reference genome\n",
            "\n",
            "Reading sample list in [/content/phasing_output_genetic_map.haps.sample]\n",
            "  * 2548 individuals included\n",
            "\n",
            "Reading site list and haplotypes in [/content/phasing_output_genetic_map.haps.haps]\n",
            "  * 98 sites included\n",
            "\n",
            "Mapping sites and samples of the haplotype set\n",
            "  * 98 sites found\n",
            "  * 0 sites not found\n",
            "  * 5096 founder haplotypes\n",
            "\n",
            "Writing haplotypes in [/content/phasing_output_genetic_map.vcf] in VCF format\n",
            "\n",
            "Running time: 1 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Part 7: what are genetic maps?\n",
        "<br>\n",
        "Heavily based on the source on part 6:\n",
        "<br>\n",
        "\n",
        "Recombination rates between SNPs are provided via a genetic map that specifies the genetic position of the SNPs in cM.\n",
        "<br>\n",
        "(so basically it contains the transition probabilities from one hidden state (haplotype) in one step to another hidden state in the next step, for all states and steps)\n",
        "\n",
        "example:\n",
        "```\n",
        "shapeit -B gwas \\\n",
        "        --input-map chr20.gmap.gz \\\n",
        "        -O gwas.phased\n",
        "```\n",
        "\n",
        " For the SNPs that don't have a genetic position, SHAPEIT internally determines its genetic position using linear interpolation.\n",
        "<br>\n",
        "\n",
        "If no genetic map is provided, SHAPEIT assumes a constant recombination rates between SNPs. This is unrealistic in human populations and thus may harm accuracy. The default value is 0.0004.\n",
        "```\n",
        "shapeit -B gwas \\\n",
        "        -O gwas.phased \\\n",
        "        --rho 0.001  # this is the constant parameter that we set\n",
        "```"
      ],
      "metadata": {
        "id": "HFAn-6sI2jvi"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Part 8\n",
        "# Compare with ground truth for one sample - genetic map used\n",
        "\n",
        "# !bcftools query -s HG00097 -f '[%CHROM\\t%POS\\t%REF\\t%ALT\\t%GT\\n]' /content/labels.vcf\n",
        "\n",
        "# just like part 5\n",
        "!whatshap compare --names truth, whatshap --tsv-pairwise eval.tsv --sample=HG00097 /content/labels.vcf /content/phasing_output_genetic_map.vcf\n",
        "\n",
        "!whatshap compare --sample=HG00097 /content/labels.vcf /content/phasing_output_genetic_map.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "U8Yu4huHuTbL",
        "outputId": "fe21f01a-dbb4-4031-c271-25beb8d3a234"
      },
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "usage: whatshap [-h] [--version] [--debug] {compare} ...\n",
            "\n",
            "positional arguments:\n",
            "  {compare}\n",
            "    compare   Compare two or more phased variant files\n",
            "\n",
            "options:\n",
            "  -h, --help  show this help message and exit\n",
            "  --version   show program's version number and exit\n",
            "  --debug     Print debug messages\n",
            "whatshap: error: unrecognized arguments: /content/labels.vcf /content/phasing_output_genetic_map.vcf\n",
            "Comparing phasings for sample HG00097\n",
            "Reading phasing from '/content/labels.vcf'\n",
            "Reading phasing from '/content/phasing_output_genetic_map.vcf'\n",
            "[W::vcf_parse] Contig '14' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
            "Chromosomes present in all VCFs: 14\n",
            "FILENAMES\n",
            "  file0 = /content/labels.vcf\n",
            "  file1 = /content/phasing_output_genetic_map.vcf\n",
            "---------------- Chromosome 14 ----------------\n",
            "VARIANT COUNTS (heterozygous / all): \n",
            "              file0:         0 /        99\n",
            "              file1:         8 /        98\n",
            "              UNION:         8 /       174\n",
            "       INTERSECTION:         0 /        23\n",
            "PAIRWISE COMPARISON: file0 <--> file1:\n",
            "         common heterozygous variants:         0\n",
            "         (restricting to these below)\n",
            "        non-singleton blocks in file0:         0\n",
            "                 --> covered variants:         0\n",
            "        non-singleton blocks in file1:         0\n",
            "                 --> covered variants:         0\n",
            "    non-singleton intersection blocks:         0\n",
            "                 --> covered variants:         0\n",
            "              ALL INTERSECTION BLOCKS: ---------\n",
            "    phased pairs of variants assessed:         0\n",
            "                        switch errors:         0\n",
            "                    switch error rate:        --\n",
            "            switch/flip decomposition:       0/0\n",
            "                     switch/flip rate:        --\n",
            "          Block-wise Hamming distance:         0\n",
            "      Block-wise Hamming distance [%]:        --\n",
            "                  Different genotypes:         0\n",
            "              Different genotypes [%]:        --\n",
            "           LARGEST INTERSECTION BLOCK: ---------\n",
            "    phased pairs of variants assessed:         0\n",
            "                        switch errors:         0\n",
            "                    switch error rate:        --\n",
            "            switch/flip decomposition:       0/0\n",
            "                     switch/flip rate:        --\n",
            "                     Hamming distance:         0\n",
            "                 Hamming distance [%]:        --\n",
            "                  Different genotypes:         0\n",
            "              Different genotypes [%]:        --\n"
          ]
        }
      ]
    }
  ]
}