{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4Bggz8oGBglV"
      },
      "source": [
        "# Computational Genomic - Practice 2\n",
        "\n",
        "## **Name:** Esra Kashaninia\n",
        "\n",
        "## **Student ID:** 402210676\n",
        "\n",
        "<br>\n",
        "\n",
        "---\n",
        "\n",
        "## Overview: Genome Assembly and Analysis Challenges\n",
        "\n",
        "In this practice, we aim to perform genome assembly and analysis for two bacteria: *E. coli* and *Shigella flexneri*. Genome as\\sembly for bacteria, despite their relatively smaller genomes, poses several challenges. Bacterial genomes can contain repetitive elements, making it difficult for short-read assemblers to resolve these repeats into single contiguous sequences. Here we utilize tools like SPAdes for assembly and BWA for read alignment, exploring metrics such as N50, total length, and contig count, which provide insights into the quality of our assembled genome.\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "b3GPTHWjXF4e"
      },
      "source": [
        "\n",
        "---\n",
        "\n",
        "## 1. Genome Assembly for *E. coli*\n",
        "\n",
        "The first step involves assembling the genome of *E. coli* using the SPAdes assembler, a popular tool designed for short-read assemblies that handles common issues such as chimeric reads and bubbles in sequencing data.\n",
        "\n",
        "<br>\n",
        "\n",
        "Step 1: Download *E. coli* Reads and Assemble with SPAdes"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "b2vvvmB_ufmS",
        "outputId": "a79e68de-900e-4cfb-97bb-3128913f2b49"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-11-27 20:17:49--  https://trace.ncbi.nlm.nih.gov/Traces/sra-reads-be/fastq?acc=SRR13921546\n",
            "Resolving trace.ncbi.nlm.nih.gov (trace.ncbi.nlm.nih.gov)... 130.14.29.113, 2607:f220:41e:4290::113\n",
            "Connecting to trace.ncbi.nlm.nih.gov (trace.ncbi.nlm.nih.gov)|130.14.29.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: unspecified [application/x-gzip]\n",
            "Saving to: ‘ecoli_reads.fastq’\n",
            "\n",
            "ecoli_reads.fastq       [              <=>   ] 363.28M  1.15MB/s    in 5m 13s  \n",
            "\n",
            "2024-11-27 20:23:04 (1.16 MB/s) - ‘ecoli_reads.fastq’ saved [380926697]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "# Download E. coli reads\n",
        "!wget -O ecoli_reads.fastq https://trace.ncbi.nlm.nih.gov/Traces/sra-reads-be/fastq?acc=SRR13921546"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4H6L0j6-uusG",
        "outputId": "dbc23ba0-b80d-4732-aa8d-02926f3e489b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  bamtools bwa libbamtools2.5.1 libhat-trie0 libhts3 libhtscodecs2 libnlopt0\n",
            "  libssw0 python3-attr python3-iniconfig python3-joblib python3-numpy\n",
            "  python3-packaging python3-pluggy python3-psutil python3-py python3-pygments\n",
            "  python3-pytest python3-simplejson python3-toml python3-yaml samtools\n",
            "Suggested packages:\n",
            "  python-attr-doc python-numpy-doc python-psutil-doc subversion\n",
            "  python-pygments-doc ttf-bitstream-vera cwltool\n",
            "The following NEW packages will be installed:\n",
            "  bamtools bwa libbamtools2.5.1 libhat-trie0 libhts3 libhtscodecs2 libnlopt0\n",
            "  libssw0 python3-attr python3-iniconfig python3-joblib python3-numpy\n",
            "  python3-packaging python3-pluggy python3-psutil python3-py python3-pygments\n",
            "  python3-pytest python3-simplejson python3-toml python3-yaml samtools spades\n",
            "0 upgraded, 23 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 11.9 MB of archives.\n",
            "After this operation, 54.8 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 python3-yaml amd64 5.4.1-1ubuntu1 [129 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libbamtools2.5.1 amd64 2.5.1+dfsg-10build1 [156 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bamtools amd64 2.5.1+dfsg-10build1 [164 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bwa amd64 0.7.17-6 [195 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhat-trie0 amd64 0.1.2-2 [11.7 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libssw0 amd64 1.1-13 [19.1 kB]\n",
            "Get:9 http://archive.ubuntu.com/ubuntu jammy/main amd64 python3-attr all 21.2.0-1 [44.0 kB]\n",
            "Get:10 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-iniconfig all 1.1.1-2 [6,024 B]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-joblib all 0.17.0-4ubuntu1 [204 kB]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 python3-numpy amd64 1:1.21.5-1ubuntu22.04.1 [3,467 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu jammy/main amd64 python3-packaging all 21.3-1 [30.7 kB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-pluggy all 0.13.0-7.1 [19.0 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy/main amd64 python3-psutil amd64 5.9.0-1build1 [158 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-py all 1.10.0-1 [71.9 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu jammy/main amd64 python3-pygments all 2.11.2+dfsg-2 [750 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-toml all 0.10.2-1 [16.5 kB]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-pytest all 6.2.5-1ubuntu2 [214 kB]\n",
            "Get:20 http://archive.ubuntu.com/ubuntu jammy/main amd64 python3-simplejson amd64 3.17.6-1build1 [54.7 kB]\n",
            "Get:21 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]\n",
            "Get:22 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libnlopt0 amd64 2.7.1-3build1 [184 kB]\n",
            "Get:23 http://archive.ubuntu.com/ubuntu jammy/universe amd64 spades amd64 3.13.1+dfsg-2ubuntu1 [5,087 kB]\n",
            "Fetched 11.9 MB in 2s (5,215 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 23.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package python3-yaml.\n",
            "(Reading database ... 123630 files and directories currently installed.)\n",
            "Preparing to unpack .../00-python3-yaml_5.4.1-1ubuntu1_amd64.deb ...\n",
            "Unpacking python3-yaml (5.4.1-1ubuntu1) ...\n",
            "Selecting previously unselected package libbamtools2.5.1:amd64.\n",
            "Preparing to unpack .../01-libbamtools2.5.1_2.5.1+dfsg-10build1_amd64.deb ...\n",
            "Unpacking libbamtools2.5.1:amd64 (2.5.1+dfsg-10build1) ...\n",
            "Selecting previously unselected package bamtools.\n",
            "Preparing to unpack .../02-bamtools_2.5.1+dfsg-10build1_amd64.deb ...\n",
            "Unpacking bamtools (2.5.1+dfsg-10build1) ...\n",
            "Selecting previously unselected package bwa.\n",
            "Preparing to unpack .../03-bwa_0.7.17-6_amd64.deb ...\n",
            "Unpacking bwa (0.7.17-6) ...\n",
            "Selecting previously unselected package libhat-trie0:amd64.\n",
            "Preparing to unpack .../04-libhat-trie0_0.1.2-2_amd64.deb ...\n",
            "Unpacking libhat-trie0:amd64 (0.1.2-2) ...\n",
            "Selecting previously unselected package libhtscodecs2:amd64.\n",
            "Preparing to unpack .../05-libhtscodecs2_1.1.1-3_amd64.deb ...\n",
            "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "Preparing to unpack .../06-libhts3_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Selecting previously unselected package libssw0:amd64.\n",
            "Preparing to unpack .../07-libssw0_1.1-13_amd64.deb ...\n",
            "Unpacking libssw0:amd64 (1.1-13) ...\n",
            "Selecting previously unselected package python3-attr.\n",
            "Preparing to unpack .../08-python3-attr_21.2.0-1_all.deb ...\n",
            "Unpacking python3-attr (21.2.0-1) ...\n",
            "Selecting previously unselected package python3-iniconfig.\n",
            "Preparing to unpack .../09-python3-iniconfig_1.1.1-2_all.deb ...\n",
            "Unpacking python3-iniconfig (1.1.1-2) ...\n",
            "Selecting previously unselected package python3-joblib.\n",
            "Preparing to unpack .../10-python3-joblib_0.17.0-4ubuntu1_all.deb ...\n",
            "Unpacking python3-joblib (0.17.0-4ubuntu1) ...\n",
            "Selecting previously unselected package python3-numpy.\n",
            "Preparing to unpack .../11-python3-numpy_1%3a1.21.5-1ubuntu22.04.1_amd64.deb ...\n",
            "Unpacking python3-numpy (1:1.21.5-1ubuntu22.04.1) ...\n",
            "Selecting previously unselected package python3-packaging.\n",
            "Preparing to unpack .../12-python3-packaging_21.3-1_all.deb ...\n",
            "Unpacking python3-packaging (21.3-1) ...\n",
            "Selecting previously unselected package python3-pluggy.\n",
            "Preparing to unpack .../13-python3-pluggy_0.13.0-7.1_all.deb ...\n",
            "Unpacking python3-pluggy (0.13.0-7.1) ...\n",
            "Selecting previously unselected package python3-psutil.\n",
            "Preparing to unpack .../14-python3-psutil_5.9.0-1build1_amd64.deb ...\n",
            "Unpacking python3-psutil (5.9.0-1build1) ...\n",
            "Selecting previously unselected package python3-py.\n",
            "Preparing to unpack .../15-python3-py_1.10.0-1_all.deb ...\n",
            "Unpacking python3-py (1.10.0-1) ...\n",
            "Selecting previously unselected package python3-pygments.\n",
            "Preparing to unpack .../16-python3-pygments_2.11.2+dfsg-2_all.deb ...\n",
            "Unpacking python3-pygments (2.11.2+dfsg-2) ...\n",
            "Selecting previously unselected package python3-toml.\n",
            "Preparing to unpack .../17-python3-toml_0.10.2-1_all.deb ...\n",
            "Unpacking python3-toml (0.10.2-1) ...\n",
            "Selecting previously unselected package python3-pytest.\n",
            "Preparing to unpack .../18-python3-pytest_6.2.5-1ubuntu2_all.deb ...\n",
            "Unpacking python3-pytest (6.2.5-1ubuntu2) ...\n",
            "Selecting previously unselected package python3-simplejson.\n",
            "Preparing to unpack .../19-python3-simplejson_3.17.6-1build1_amd64.deb ...\n",
            "Unpacking python3-simplejson (3.17.6-1build1) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../20-samtools_1.13-4_amd64.deb ...\n",
            "Unpacking samtools (1.13-4) ...\n",
            "Selecting previously unselected package libnlopt0:amd64.\n",
            "Preparing to unpack .../21-libnlopt0_2.7.1-3build1_amd64.deb ...\n",
            "Unpacking libnlopt0:amd64 (2.7.1-3build1) ...\n",
            "Selecting previously unselected package spades.\n",
            "Preparing to unpack .../22-spades_3.13.1+dfsg-2ubuntu1_amd64.deb ...\n",
            "Unpacking spades (3.13.1+dfsg-2ubuntu1) ...\n",
            "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Setting up python3-iniconfig (1.1.1-2) ...\n",
            "Setting up python3-attr (21.2.0-1) ...\n",
            "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Setting up python3-py (1.10.0-1) ...\n",
            "Setting up python3-joblib (0.17.0-4ubuntu1) ...\n",
            "Setting up bwa (0.7.17-6) ...\n",
            "Setting up python3-yaml (5.4.1-1ubuntu1) ...\n",
            "Setting up libnlopt0:amd64 (2.7.1-3build1) ...\n",
            "Setting up python3-psutil (5.9.0-1build1) ...\n",
            "Setting up samtools (1.13-4) ...\n",
            "Setting up python3-simplejson (3.17.6-1build1) ...\n",
            "Setting up python3-pygments (2.11.2+dfsg-2) ...\n",
            "Setting up python3-packaging (21.3-1) ...\n",
            "Setting up libhat-trie0:amd64 (0.1.2-2) ...\n",
            "Setting up libbamtools2.5.1:amd64 (2.5.1+dfsg-10build1) ...\n",
            "Setting up python3-numpy (1:1.21.5-1ubuntu22.04.1) ...\n",
            "Setting up python3-toml (0.10.2-1) ...\n",
            "Setting up python3-pluggy (0.13.0-7.1) ...\n",
            "Setting up bamtools (2.5.1+dfsg-10build1) ...\n",
            "Setting up libssw0:amd64 (1.1-13) ...\n",
            "Setting up python3-pytest (6.2.5-1ubuntu2) ...\n",
            "Setting up spades (3.13.1+dfsg-2ubuntu1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n"
          ]
        }
      ],
      "source": [
        "# Install SPAdes if not already installed\n",
        "!sudo apt-get install spades"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BHAKeLqDuxdF",
        "outputId": "3c06f758-60c7-402e-ab1c-6ffb90a3c3cd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Command line: /usr/lib/spades/bin/spades.py\t-s\t/content/ecoli_reads.fastq\t-o\t/content/spade_ecoli\t--only-assembler\t-k\t21,33,55\t\n",
            "\n",
            "System information:\n",
            "  SPAdes version: 3.13.1\n",
            "  Python version: 3.10.12\n",
            "  OS: Linux-6.1.85+-x86_64-with-glibc2.35\n",
            "\n",
            "Output dir: /content/spade_ecoli\n",
            "Mode: ONLY assembling (without read error correction)\n",
            "Debug mode is turned OFF\n",
            "\n",
            "Dataset parameters:\n",
            "  Multi-cell mode (you should set '--sc' flag if input data was obtained with MDA (single-cell) technology or --meta flag if processing metagenomic dataset)\n",
            "  Reads:\n",
            "    Library number: 1, library type: single\n",
            "      left reads: not specified\n",
            "      right reads: not specified\n",
            "      interlaced reads: not specified\n",
            "      single reads: ['/content/ecoli_reads.fastq']\n",
            "      merged reads: not specified\n",
            "Assembly parameters:\n",
            "  k: [21, 33, 55]\n",
            "  Repeat resolution is enabled\n",
            "  Mismatch careful mode is turned OFF\n",
            "  MismatchCorrector will be SKIPPED\n",
            "  Coverage cutoff is turned OFF\n",
            "Other parameters:\n",
            "  Dir for temp files: /content/spade_ecoli/tmp\n",
            "  Threads: 16\n",
            "  Memory limit (in Gb): 12\n",
            "\n",
            "\n",
            "======= SPAdes pipeline started. Log can be found here: /content/spade_ecoli/spades.log\n",
            "\n",
            "\n",
            "===== Assembling started.\n",
            "\n",
            "\n",
            "== Running assembler: K21\n",
            "\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /content/spade_ecoli/K21/configs/config.info\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 12 Gb\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  84)   Starting SPAdes, built from N/A, git revision N/A\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/content/spade_ecoli/dataset.info) with K=21\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 2\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  52)   SPAdes started\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  59)   Starting from stage: construction\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  66)   Two-step RR enabled: 0\n",
            "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  77)   Converting reads to binary format for library #0 (takes a while)\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  78)   Converting paired reads\n",
            "  0:00:00.092    68M / 100M  INFO    General                 (binary_converter.hpp      : 117)   0 reads written\n",
            "  0:00:00.096     4M / 100M  INFO    General                 (read_converter.hpp        :  87)   Converting single reads\n",
            "  0:00:00.309   136M / 196M  INFO    General                 (binary_converter.hpp      :  93)   16384 reads processed\n",
            "  0:00:00.355   140M / 196M  INFO    General                 (binary_converter.hpp      :  93)   32768 reads processed\n",
            "  0:00:00.448   148M / 196M  INFO    General                 (binary_converter.hpp      :  93)   65536 reads processed\n",
            "  0:00:00.646   164M / 196M  INFO    General                 (binary_converter.hpp      :  93)   131072 reads processed\n",
            "  0:00:01.049   200M / 200M  INFO    General                 (binary_converter.hpp      :  93)   262144 reads processed\n",
            "  0:00:01.862   268M / 268M  INFO    General                 (binary_converter.hpp      :  93)   524288 reads processed\n",
            "  0:00:03.507   408M / 408M  INFO    General                 (binary_converter.hpp      :  93)   1048576 reads processed\n",
            "  0:00:09.446   480M / 480M  INFO    General                 (binary_converter.hpp      :  93)   2097152 reads processed\n",
            "  0:00:20.686   512M / 512M  INFO    General                 (binary_converter.hpp      :  93)   4194304 reads processed\n",
            "  0:00:39.724   512M / 512M  INFO    General                 (binary_converter.hpp      :  93)   8388608 reads processed\n",
            "  0:00:47.978   328M / 512M  INFO    General                 (binary_converter.hpp      : 117)   9703580 reads written\n",
            "  0:00:48.107     4M / 512M  INFO    General                 (read_converter.hpp        :  95)   Converting merged reads\n",
            "  0:00:48.308   132M / 512M  INFO    General                 (binary_converter.hpp      : 117)   0 reads written\n",
            "  0:00:48.317     4M / 512M  INFO    General                 (construction.cpp          : 111)   Max read length 100\n",
            "  0:00:48.317     4M / 512M  INFO    General                 (construction.cpp          : 117)   Average read length 99.9994\n",
            "  0:00:48.317     4M / 512M  INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
            "  0:00:48.317     4M / 512M  INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 4 files using 2 threads. This might take a while.\n",
            "  0:00:48.317     4M / 512M  INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1048576\n",
            "  0:00:48.317     4M / 512M  INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 1.99935 Gb\n",
            "  0:00:48.317     4M / 512M  INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 16777216\n",
            "  0:01:17.352     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 3392280 reads\n",
            "  0:01:45.573     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 6784678 reads\n",
            "  0:02:13.705     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 10177466 reads\n",
            "  0:02:43.199     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 13569862 reads\n",
            "  0:03:13.938     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 16962736 reads\n",
            "  0:03:35.414     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 19407160 reads\n",
            "  0:03:35.414     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
            "  0:03:35.585     8M / 1G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 19407160 reads\n",
            "  0:03:35.585     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
            "  0:03:38.597     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 20927854 kmers in total.\n",
            "  0:03:38.597     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
            "  0:03:38.949     8M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
            "  0:03:38.949     8M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
            "  0:03:38.949     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 32 files using 2 threads. This might take a while.\n",
            "  0:03:38.949     8M / 1G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1048576\n",
            "  0:03:38.949     8M / 1G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 1.9987 Gb\n",
            "  0:03:38.949     8M / 1G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 2097152\n",
            "  0:03:49.689     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 20927854 kmers\n",
            "  0:03:49.689     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 20927854 kmers.\n",
            "  0:03:49.747     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
            "  0:03:51.039     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 20555461 kmers in total.\n",
            "  0:03:51.040     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
            "  0:03:51.353     8M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
            "  0:03:54.697    16M / 1G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
            "  0:03:55.016    16M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 9539944 bytes occupied (3.71286 bits per kmer).\n",
            "  0:03:55.030    36M / 1G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
            "  0:04:07.800    36M / 1G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
            "  0:04:07.801    36M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
            "  0:04:07.801    36M / 1G    INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
            "  0:04:07.801    36M / 1G    INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
            "  0:04:35.991    36M / 1G    INFO   Early tip clipping       (early_simplification.hpp  : 184)   5852138 22-mers were removed by early tip clipper\n",
            "  0:04:35.991    36M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
            "  0:04:35.991    36M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
            "  0:04:51.055    96M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 1130550 sequences extracted\n",
            "  0:04:58.868    96M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
            "  0:05:01.722    96M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 1 loops collected\n",
            "  0:05:03.641   344M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
            "  0:05:03.641   344M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
            "  0:05:03.641   344M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
            "  0:05:11.493   352M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 9705264 bytes occupied (3.70999 bits per kmer).\n",
            "  0:05:11.562   432M / 1G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
            "  0:09:27.886   432M / 1G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
            "  0:09:37.366   432M / 1G    INFO    General                 (construction.cpp          : 464)   Processed 2000000 edges\n",
            "  0:09:38.361   432M / 1G    INFO    General                 (construction.cpp          : 464)   Processed 2260962 edges\n",
            "  0:09:38.448   316M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
            "  0:09:38.449   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 21\n",
            "  0:09:38.449   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 100\n",
            "  0:09:38.449   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 120. Coverage mad: 51.891\n",
            "  0:09:38.449   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
            "  0:09:38.700   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
            "  0:09:39.524   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
            "  0:09:42.473   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
            "  0:09:49.064   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
            "  0:09:53.444   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 155.923. Fitted coverage std. dev: 58.3447\n",
            "  0:09:53.450   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.906432\n",
            "  0:09:53.450   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 37\n",
            "  0:09:53.450   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 37\n",
            "  0:09:53.450   316M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 4591618\n",
            "  0:09:53.450   316M / 1G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 155.923\n",
            "  0:09:53.450   316M / 1G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 37\n",
            "  0:09:53.450   316M / 1G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
            "  0:09:53.450   316M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
            "  0:09:53.450   316M / 1G    INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
            "  0:09:53.451   316M / 1G    INFO    General                 (graph_simplification.hpp  : 669)   Flanking coverage based disconnection disabled\n",
            "  0:09:53.451   316M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
            "  0:09:53.848   316M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
            "  0:09:53.848   316M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial tip clipper\n",
            "  0:09:54.650   316M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial tip clipper triggered 4781 times\n",
            "  0:09:54.650   316M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial ec remover\n",
            "  0:10:12.750   344M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial ec remover triggered 344718 times\n",
            "  0:10:12.751   344M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial isolated edge remover\n",
            "  0:10:13.226   344M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial isolated edge remover triggered 20867 times\n",
            "  0:10:13.344   328M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
            "  0:10:13.344   328M / 1G    INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
            "  0:10:13.344   328M / 1G    INFO    General                 (graph_simplification.hpp  : 641)   Creating parallel br instance\n",
            "  0:10:13.344   328M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
            "  0:10:13.344   328M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:13.707   324M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 2374 times\n",
            "  0:10:13.707   324M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:20.760   204M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 38920 times\n",
            "  0:10:20.760   204M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.165   128M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 6057 times\n",
            "  0:10:21.165   128M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
            "  0:10:21.165   128M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.172   124M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 79 times\n",
            "  0:10:21.172   124M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.278   116M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 182 times\n",
            "  0:10:21.278   116M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.381    88M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1335 times\n",
            "  0:10:21.381    88M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
            "  0:10:21.381    88M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.382    88M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 8 times\n",
            "  0:10:21.382    88M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.412    84M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 48 times\n",
            "  0:10:21.412    84M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.435    76M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 366 times\n",
            "  0:10:21.435    76M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
            "  0:10:21.435    76M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.442    76M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:10:21.442    76M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.447    72M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 18 times\n",
            "  0:10:21.447    72M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.461    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 190 times\n",
            "  0:10:21.461    68M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
            "  0:10:21.461    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.461    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:10:21.461    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.470    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 9 times\n",
            "  0:10:21.470    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.475    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 86 times\n",
            "  0:10:21.475    68M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
            "  0:10:21.475    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.475    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:10:21.475    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.476    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 2 times\n",
            "  0:10:21.476    68M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.478    64M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 47 times\n",
            "  0:10:21.478    64M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
            "  0:10:21.478    64M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.478    64M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:10:21.478    64M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.479    64M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 2 times\n",
            "  0:10:21.479    64M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.483    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 66 times\n",
            "  0:10:21.483    60M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
            "  0:10:21.483    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.484    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:10:21.484    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.491    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 3 times\n",
            "  0:10:21.491    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.494    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 35 times\n",
            "  0:10:21.494    60M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
            "  0:10:21.494    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.494    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:10:21.494    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.494    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:10:21.494    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.504    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 42 times\n",
            "  0:10:21.504    60M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
            "  0:10:21.504    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.504    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:10:21.504    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.505    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
            "  0:10:21.505    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.506    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 32 times\n",
            "  0:10:21.506    60M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
            "  0:10:21.506    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.508    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:10:21.508    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.525    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:10:21.525    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:10:21.528    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
            "  0:10:21.534    56M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
            "  0:10:21.534    56M / 1G    INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
            "  0:10:21.534    56M / 1G    INFO    General                 (graph_simplification.hpp  : 458)   Disconnection of relatively low covered edges disabled\n",
            "  0:10:21.534    56M / 1G    INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
            "  0:10:21.534    56M / 1G    INFO    General                 (graph_simplification.hpp  : 641)   Creating parallel br instance\n",
            "  0:10:21.534    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.536    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:10:21.536    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.552    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:10:21.552    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:10:21.554    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:10:21.554    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:10:21.569    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:10:21.569    56M / 1G    INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
            "  0:10:21.573    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
            "  0:10:21.575    56M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 76 times\n",
            "  0:10:21.575    56M / 1G    INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
            "  0:10:21.576    56M / 1G    INFO    General                 (simplification.cpp        : 476)   Average coverage = 163.331\n",
            "  0:10:21.576    56M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
            "  0:10:21.576    56M / 1G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /content/spade_ecoli//K21/simplified_contigs.fasta\n",
            "  0:10:21.651    56M / 1G    INFO    General                 (launch.hpp                : 151)   SPAdes finished\n",
            "  0:10:21.720    12M / 1G    INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 10 minutes 21 seconds\n",
            "Max read length detected as 100\n",
            "\n",
            "== Running assembler: K33\n",
            "\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /content/spade_ecoli/K33/configs/config.info\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 12 Gb\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  84)   Starting SPAdes, built from N/A, git revision N/A\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/content/spade_ecoli/dataset.info) with K=33\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 2\n",
            "  0:00:00.001     4M / 4M    INFO    General                 (launch.hpp                :  52)   SPAdes started\n",
            "  0:00:00.001     4M / 4M    INFO    General                 (launch.hpp                :  59)   Starting from stage: construction\n",
            "  0:00:00.001     4M / 4M    INFO    General                 (launch.hpp                :  66)   Two-step RR enabled: 0\n",
            "  0:00:00.001     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
            "  0:00:00.001     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
            "  0:00:00.002     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 100\n",
            "  0:00:00.002     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 99.9994\n",
            "  0:00:00.002     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
            "  0:00:00.002     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 4 files using 2 threads. This might take a while.\n",
            "  0:00:00.002     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1048576\n",
            "  0:00:00.002     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 1.99935 Gb\n",
            "  0:00:00.002     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 8388608\n",
            "  0:00:16.270     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 2000083 reads\n",
            "  0:00:35.072     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 4000055 reads\n",
            "  0:00:50.695     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 5999858 reads\n",
            "  0:01:06.535     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 8000176 reads\n",
            "  0:01:22.197     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 10000448 reads\n",
            "  0:01:37.913     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 12001519 reads\n",
            "  0:01:53.496     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 14001585 reads\n",
            "  0:02:09.623     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 16002638 reads\n",
            "  0:02:26.229     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 18003661 reads\n",
            "  0:02:36.715     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 19407160 reads\n",
            "  0:02:36.715     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
            "  0:02:39.007     8M / 1G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 19407160 reads\n",
            "  0:02:39.007     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
            "  0:02:45.133     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 23563111 kmers in total.\n",
            "  0:02:45.134     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
            "  0:02:47.171     8M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
            "  0:02:47.172     8M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
            "  0:02:47.172     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 32 files using 2 threads. This might take a while.\n",
            "  0:02:47.173     8M / 1G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1048576\n",
            "  0:02:47.173     8M / 1G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 1.9987 Gb\n",
            "  0:02:47.173     8M / 1G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 1048576\n",
            "  0:03:07.094     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 23563111 kmers\n",
            "  0:03:07.094     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 23563111 kmers.\n",
            "  0:03:07.258     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
            "  0:03:10.486     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 23447819 kmers in total.\n",
            "  0:03:10.487     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
            "  0:03:13.893     8M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
            "  0:03:17.814    16M / 1G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
            "  0:03:18.668    16M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 10881024 bytes occupied (3.71242 bits per kmer).\n",
            "  0:03:18.688    40M / 1G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
            "  0:03:33.378    40M / 1G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
            "  0:03:33.378    40M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
            "  0:03:33.378    40M / 1G    INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
            "  0:03:33.378    40M / 1G    INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
            "  0:04:00.771    40M / 1G    INFO   Early tip clipping       (early_simplification.hpp  : 184)   9445044 34-mers were removed by early tip clipper\n",
            "  0:04:00.771    40M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
            "  0:04:00.772    40M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
            "  0:04:15.170    76M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 554917 sequences extracted\n",
            "  0:04:22.213    76M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
            "  0:04:25.082    76M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 2 loops collected\n",
            "  0:04:26.162   204M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
            "  0:04:26.162   204M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
            "  0:04:26.162   204M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
            "  0:04:35.857   216M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 10927240 bytes occupied (3.70995 bits per kmer).\n",
            "  0:04:35.922   308M / 1G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
            "  0:08:13.414   308M / 1G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
            "  0:08:22.004   308M / 1G    INFO    General                 (construction.cpp          : 464)   Processed 1109816 edges\n",
            "  0:08:22.233   172M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
            "  0:08:22.235   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 15\n",
            "  0:08:22.235   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 85\n",
            "  0:08:22.236   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 103. Coverage mad: 45.9606\n",
            "  0:08:22.236   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
            "  0:08:22.567   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
            "  0:08:23.478   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
            "  0:08:26.254   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
            "  0:08:31.918   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
            "  0:08:34.997   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 131.277. Fitted coverage std. dev: 49.4795\n",
            "  0:08:35.004   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.876065\n",
            "  0:08:35.005   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 27\n",
            "  0:08:35.005   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 27\n",
            "  0:08:35.005   172M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 4606203\n",
            "  0:08:35.005   172M / 1G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 131.277\n",
            "  0:08:35.005   172M / 1G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 27\n",
            "  0:08:35.005   172M / 1G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
            "  0:08:35.005   172M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
            "  0:08:35.005   172M / 1G    INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
            "  0:08:35.005   172M / 1G    INFO    General                 (graph_simplification.hpp  : 669)   Flanking coverage based disconnection disabled\n",
            "  0:08:35.006   172M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
            "  0:08:35.314   172M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
            "  0:08:35.314   172M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial tip clipper\n",
            "  0:08:36.028   168M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial tip clipper triggered 8266 times\n",
            "  0:08:36.028   168M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial ec remover\n",
            "  0:08:42.921   184M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial ec remover triggered 152952 times\n",
            "  0:08:42.921   184M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial isolated edge remover\n",
            "  0:08:43.209   176M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial isolated edge remover triggered 51966 times\n",
            "  0:08:43.235   168M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
            "  0:08:43.235   168M / 1G    INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
            "  0:08:43.235   168M / 1G    INFO    General                 (graph_simplification.hpp  : 641)   Creating parallel br instance\n",
            "  0:08:43.235   168M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
            "  0:08:43.235   168M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:43.304   164M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1910 times\n",
            "  0:08:43.304   164M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.335    88M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 11776 times\n",
            "  0:08:44.335    88M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.486    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1368 times\n",
            "  0:08:44.486    60M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
            "  0:08:44.486    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.488    52M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 30 times\n",
            "  0:08:44.488    52M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.500    52M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 37 times\n",
            "  0:08:44.500    52M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.578    40M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 563 times\n",
            "  0:08:44.578    40M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
            "  0:08:44.578    40M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.579    40M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 8 times\n",
            "  0:08:44.579    40M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.583    36M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 24 times\n",
            "  0:08:44.583    36M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.595    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 71 times\n",
            "  0:08:44.595    32M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
            "  0:08:44.595    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.595    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:08:44.595    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.596    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 6 times\n",
            "  0:08:44.596    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.602    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 40 times\n",
            "  0:08:44.602    32M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
            "  0:08:44.602    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.603    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:08:44.603    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.603    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
            "  0:08:44.603    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.604    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 6 times\n",
            "  0:08:44.604    32M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
            "  0:08:44.604    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.606    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:08:44.606    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.609    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 2 times\n",
            "  0:08:44.609    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.609    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 4 times\n",
            "  0:08:44.609    32M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
            "  0:08:44.609    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.609    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:08:44.609    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.609    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
            "  0:08:44.609    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.611    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 7 times\n",
            "  0:08:44.611    32M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
            "  0:08:44.611    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.611    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:08:44.611    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.611    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:08:44.611    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.612    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 4 times\n",
            "  0:08:44.612    32M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
            "  0:08:44.612    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.612    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:08:44.612    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.612    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:08:44.612    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.613    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 6 times\n",
            "  0:08:44.613    32M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
            "  0:08:44.613    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.613    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:08:44.613    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.613    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:08:44.613    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.614    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 4 times\n",
            "  0:08:44.614    32M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
            "  0:08:44.614    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.615    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:08:44.615    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.618    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:08:44.618    32M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:08:44.630    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
            "  0:08:44.630    28M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
            "  0:08:44.630    28M / 1G    INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
            "  0:08:44.630    28M / 1G    INFO    General                 (graph_simplification.hpp  : 458)   Disconnection of relatively low covered edges disabled\n",
            "  0:08:44.630    28M / 1G    INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
            "  0:08:44.630    28M / 1G    INFO    General                 (graph_simplification.hpp  : 641)   Creating parallel br instance\n",
            "  0:08:44.630    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.631    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:08:44.631    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.635    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:08:44.635    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:08:44.636    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:08:44.636    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:08:44.639    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:08:44.640    28M / 1G    INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
            "  0:08:44.641    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
            "  0:08:44.642    28M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 74 times\n",
            "  0:08:44.642    28M / 1G    INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
            "  0:08:44.642    28M / 1G    INFO    General                 (simplification.cpp        : 476)   Average coverage = 136.914\n",
            "  0:08:44.642    28M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
            "  0:08:44.643    28M / 1G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /content/spade_ecoli//K33/simplified_contigs.fasta\n",
            "  0:08:44.711    28M / 1G    INFO    General                 (launch.hpp                : 151)   SPAdes finished\n",
            "  0:08:44.733    12M / 1G    INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 8 minutes 44 seconds\n",
            "\n",
            "== Running assembler: K55\n",
            "\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /content/spade_ecoli/K55/configs/config.info\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 12 Gb\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  84)   Starting SPAdes, built from N/A, git revision N/A\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/content/spade_ecoli/dataset.info) with K=55\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 2\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  52)   SPAdes started\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  59)   Starting from stage: construction\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  66)   Two-step RR enabled: 0\n",
            "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  77)   Will need read mapping, kmer mapper will be attached\n",
            "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
            "  0:00:00.005     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
            "  0:00:00.005     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 100\n",
            "  0:00:00.005     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 99.9994\n",
            "  0:00:00.005     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
            "  0:00:00.006     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 4 files using 2 threads. This might take a while.\n",
            "  0:00:00.006     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1048576\n",
            "  0:00:00.006     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 1.99935 Gb\n",
            "  0:00:00.006     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 8388608\n",
            "  0:00:17.134     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 2979839 reads\n",
            "  0:00:34.581     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 5958351 reads\n",
            "  0:00:50.861     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 8938194 reads\n",
            "  0:01:07.036     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 11917467 reads\n",
            "  0:01:23.192     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 14896240 reads\n",
            "  0:01:39.510     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 17874956 reads\n",
            "  0:01:48.703     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 19407160 reads\n",
            "  0:01:48.704     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
            "  0:01:50.275     8M / 1G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 19407160 reads\n",
            "  0:01:50.275     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
            "  0:01:54.805     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 22660742 kmers in total.\n",
            "  0:01:54.805     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
            "  0:01:56.742     8M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
            "  0:01:56.743     8M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
            "  0:01:56.743     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 32 files using 2 threads. This might take a while.\n",
            "  0:01:56.743     8M / 1G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1048576\n",
            "  0:01:56.743     8M / 1G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 1.9987 Gb\n",
            "  0:01:56.743     8M / 1G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 1048576\n",
            "  0:02:09.093     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 22660742 kmers\n",
            "  0:02:09.093     1G / 1G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 22660742 kmers.\n",
            "  0:02:09.280     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
            "  0:02:14.164     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 22824723 kmers in total.\n",
            "  0:02:14.164     8M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
            "  0:02:15.141     8M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
            "  0:02:18.735    16M / 1G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
            "  0:02:19.485    16M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 10591952 bytes occupied (3.71245 bits per kmer).\n",
            "  0:02:19.500    40M / 1G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
            "  0:02:32.802    40M / 1G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
            "  0:02:32.802    40M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
            "  0:02:32.802    40M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
            "  0:02:53.010   120M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 1193139 sequences extracted\n",
            "  0:03:02.674   120M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
            "  0:03:06.374   120M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 0 loops collected\n",
            "  0:03:08.805   432M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
            "  0:03:08.806   432M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
            "  0:03:08.806   432M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
            "  0:03:16.363   444M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 10508840 bytes occupied (3.70997 bits per kmer).\n",
            "  0:03:16.446   532M / 1G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
            "  0:06:01.635   532M / 1G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
            "  0:06:16.209   532M / 1G    INFO    General                 (construction.cpp          : 464)   Processed 2000000 edges\n",
            "  0:06:19.773   532M / 1G    INFO    General                 (construction.cpp          : 464)   Processed 2386276 edges\n",
            "  0:06:19.917   400M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
            "  0:06:19.917   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 11\n",
            "  0:06:19.917   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 56\n",
            "  0:06:19.918   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 68. Coverage mad: 31.1346\n",
            "  0:06:19.918   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
            "  0:06:20.056   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
            "  0:06:20.502   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
            "  0:06:22.217   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
            "  0:06:26.061   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
            "  0:06:27.342   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 87.0937. Fitted coverage std. dev: 33.4839\n",
            "  0:06:27.346   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.499988\n",
            "  0:06:27.346   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 17\n",
            "  0:06:27.346   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 17\n",
            "  0:06:27.346   400M / 1G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 4616054\n",
            "  0:06:27.346   400M / 1G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 87.0937\n",
            "  0:06:27.346   400M / 1G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 17\n",
            "  0:06:27.346   400M / 1G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
            "  0:06:27.346   400M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Gap Closer\n",
            "  0:06:27.346   400M / 1G    INFO    General                 (gap_closer.cpp            : 427)   No paired-end libraries exist, skipping gap closer\n",
            "  0:06:27.346   400M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
            "  0:06:27.346   400M / 1G    INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
            "  0:06:27.346   400M / 1G    INFO    General                 (graph_simplification.hpp  : 669)   Flanking coverage based disconnection disabled\n",
            "  0:06:27.346   400M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
            "  0:06:27.688   400M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
            "  0:06:27.688   400M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial tip clipper\n",
            "  0:06:40.742   400M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial tip clipper triggered 522239 times\n",
            "  0:06:40.742   400M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial ec remover\n",
            "  0:06:41.137   384M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial ec remover triggered 2598 times\n",
            "  0:06:41.137   384M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial isolated edge remover\n",
            "  0:06:42.636    60M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial isolated edge remover triggered 167310 times\n",
            "  0:06:42.653    52M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
            "  0:06:42.653    52M / 1G    INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
            "  0:06:42.653    52M / 1G    INFO    General                 (graph_simplification.hpp  : 641)   Creating parallel br instance\n",
            "  0:06:42.654    52M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
            "  0:06:42.654    52M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:42.708    44M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 171 times\n",
            "  0:06:42.708    44M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.129    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 524 times\n",
            "  0:06:43.129    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.138    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
            "  0:06:43.138    24M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
            "  0:06:43.139    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.139    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:06:43.139    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.139    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:06:43.139    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.168    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 21 times\n",
            "  0:06:43.169    24M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
            "  0:06:43.169    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.169    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
            "  0:06:43.169    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.187    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 11 times\n",
            "  0:06:43.188    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.193    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 12 times\n",
            "  0:06:43.193    24M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
            "  0:06:43.193    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.194    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 5 times\n",
            "  0:06:43.195    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.195    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
            "  0:06:43.196    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.196    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
            "  0:06:43.196    24M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
            "  0:06:43.196    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.196    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:06:43.196    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.196    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:06:43.196    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.198    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
            "  0:06:43.198    24M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
            "  0:06:43.198    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.199    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:06:43.199    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.199    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:06:43.199    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.199    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
            "  0:06:43.199    24M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
            "  0:06:43.199    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.199    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 2 times\n",
            "  0:06:43.199    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.199    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:06:43.199    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.200    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
            "  0:06:43.200    24M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
            "  0:06:43.200    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.200    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 2 times\n",
            "  0:06:43.200    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.200    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:06:43.200    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.200    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
            "  0:06:43.200    24M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
            "  0:06:43.200    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.201    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:06:43.201    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.201    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:06:43.201    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.201    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
            "  0:06:43.201    24M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
            "  0:06:43.202    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.202    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:06:43.202    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.202    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:06:43.202    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.202    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
            "  0:06:43.202    24M / 1G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
            "  0:06:43.202    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.209    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:06:43.209    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.220    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:06:43.221    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
            "  0:06:43.233    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
            "  0:06:43.233    24M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Gap Closer\n",
            "  0:06:43.234    24M / 1G    INFO    General                 (gap_closer.cpp            : 427)   No paired-end libraries exist, skipping gap closer\n",
            "  0:06:43.234    24M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
            "  0:06:43.234    24M / 1G    INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
            "  0:06:43.234    24M / 1G    INFO    General                 (graph_simplification.hpp  : 458)   Disconnection of relatively low covered edges disabled\n",
            "  0:06:43.234    24M / 1G    INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
            "  0:06:43.234    24M / 1G    INFO    General                 (graph_simplification.hpp  : 641)   Creating parallel br instance\n",
            "  0:06:43.234    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.249    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:06:43.249    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.261    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:06:43.262    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
            "  0:06:43.279    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
            "  0:06:43.279    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
            "  0:06:43.297    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
            "  0:06:43.297    24M / 1G    INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
            "  0:06:43.300    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
            "  0:06:43.316    24M / 1G    INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 3 times\n",
            "  0:06:43.316    24M / 1G    INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
            "  0:06:43.317    24M / 1G    INFO    General                 (simplification.cpp        : 476)   Average coverage = 90.4869\n",
            "  0:06:43.317    24M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Mismatch Correction\n",
            "  0:06:43.317    24M / 1G    INFO    General                 (graph_pack.hpp            : 101)   Index refill\n",
            "  0:06:43.318    24M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
            "  0:06:43.318    24M / 1G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 32 files using 2 threads. This might take a while.\n",
            "  0:06:43.318    24M / 1G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1048576\n",
            "  0:06:43.318    24M / 1G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 1.99609 Gb\n",
            "  0:06:43.318    24M / 1G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 1048576\n",
            "  0:06:44.779     1G / 1G    INFO    General                 (edge_index_builders.hpp   :  82)   Used 930 sequences.\n",
            "  0:06:44.796    24M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
            "  0:06:45.365    24M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 4619512 kmers in total.\n",
            "  0:06:45.365    24M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
            "  0:06:45.467    24M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
            "  0:06:46.240    24M / 1G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
            "  0:06:46.426    24M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 2150456 bytes occupied (3.72413 bits per kmer).\n",
            "  0:06:46.590   132M / 1G    INFO    General                 (edge_index_builders.hpp   : 107)   Collecting edge information from graph, this takes a while.\n",
            "  0:06:47.763   132M / 1G    INFO    General                 (edge_index.hpp            :  92)   Index refilled\n",
            "  0:06:47.763   132M / 1G    INFO    General                 (graph_pack.hpp            : 109)   Normalizing k-mer map. Total 64708 kmers to process\n",
            "  0:06:47.828   132M / 1G    INFO    General                 (graph_pack.hpp            : 111)   Normalizing done\n",
            "  0:08:33.477   132M / 1G    INFO    General                 (mismatch_shall_not_pass.hp: 189)   Finished collecting potential mismatches positions\n",
            "  0:08:33.561   132M / 1G    INFO    General                 (mismatch_shall_not_pass.hp: 290)   All edges processed\n",
            "  0:08:33.562   132M / 1G    INFO    General                 (mismatch_correction.cpp   :  27)   Corrected 1 nucleotides\n",
            "  0:08:33.562   132M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
            "  0:08:33.562   132M / 1G    INFO    General                 (contig_output_stage.cpp   :  45)   Writing GFA to /content/spade_ecoli//K55/assembly_graph_with_scaffolds.gfa\n",
            "  0:08:33.631   132M / 1G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /content/spade_ecoli//K55/before_rr.fasta\n",
            "  0:08:33.715   132M / 1G    INFO    General                 (contig_output_stage.cpp   :  56)   Outputting FastG graph to /content/spade_ecoli//K55/assembly_graph.fastg\n",
            "  0:08:33.955   132M / 1G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /content/spade_ecoli//K55/final_contigs.fasta\n",
            "  0:08:34.034   132M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Paired Information Counting\n",
            "  0:08:34.037   132M / 1G    INFO    General                 (graph_pack.hpp            : 109)   Normalizing k-mer map. Total 64734 kmers to process\n",
            "  0:08:34.125   132M / 1G    INFO    General                 (graph_pack.hpp            : 111)   Normalizing done\n",
            "  0:08:34.126   132M / 1G    INFO    General                 (pair_info_count.cpp       : 323)   Min edge length for estimation: 60713\n",
            "  0:08:34.126   132M / 1G    INFO    General                 (pair_info_count.cpp       : 395)   Mapping single reads of library #0\n",
            "  0:08:34.126   132M / 1G    INFO    General                 (pair_info_count.cpp       :  39)   Selecting usual mapper\n",
            "  0:08:35.825   132M / 1G    INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 200000 reads\n",
            "  0:08:36.160   132M / 1G    INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 400000 reads\n",
            "  0:08:37.387   132M / 1G    INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 600000 reads\n",
            "  0:08:38.115   132M / 1G    INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 800000 reads\n",
            "  0:08:38.974   132M / 1G    INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1000000 reads\n",
            "  0:08:40.099   132M / 1G    INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1200000 reads\n",
            "  0:08:43.966   136M / 1G    INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 2200000 reads\n",
            "  0:08:54.922   132M / 1G    INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 4200000 reads\n",
            "  0:09:16.987   132M / 1G    INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 8400000 reads\n",
            "  0:09:26.661   132M / 1G    INFO    General                 (sequence_mapper_notifier.h:  98)   Total 9703580 reads processed\n",
            "  0:09:26.661   132M / 1G    INFO    General                 (pair_info_count.cpp       : 397)   Total paths obtained from single reads: 2734\n",
            "  0:09:26.661   132M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Distance Estimation\n",
            "  0:09:26.661   132M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Repeat Resolving\n",
            "  0:09:26.661   132M / 1G    INFO    General                 (repeat_resolving.cpp      :  69)   Using Path-Extend repeat resolving\n",
            "  0:09:26.661   132M / 1G    INFO    General                 (launcher.cpp              : 481)   ExSPAnder repeat resolving tool started\n",
            "  0:09:26.665   136M / 1G    INFO    General                 (launcher.cpp              : 392)   Creating main extenders, unique edge length = 2000\n",
            "  0:09:26.688   144M / 1G    INFO    General                 (extenders_logic.cpp       :  34)   resolvable_repeat_length_bound set to 10000\n",
            "  0:09:26.699   144M / 1G    INFO    General                 (extenders_logic.cpp       : 475)   Using 0 paired-end libraries\n",
            "  0:09:26.700   144M / 1G    INFO    General                 (extenders_logic.cpp       : 476)   Using 0 paired-end scaffolding libraries\n",
            "  0:09:26.700   144M / 1G    INFO    General                 (extenders_logic.cpp       : 477)   Using 1 single read library\n",
            "  0:09:26.700   144M / 1G    INFO    General                 (launcher.cpp              : 420)   Total number of extenders is 1\n",
            "  0:09:26.700   144M / 1G    INFO    General                 (path_extender.hpp         : 885)   Processed 0 paths from 420 (0%)\n",
            "  0:09:26.716   144M / 1G    INFO    General                 (path_extender.hpp         : 885)   Processed 43 paths from 420 (10%)\n",
            "  0:09:26.728   148M / 1G    INFO    General                 (path_extender.hpp         : 885)   Processed 86 paths from 420 (20%)\n",
            "  0:09:26.735   148M / 1G    INFO    General                 (path_extender.hpp         : 883)   Processed 128 paths from 420 (30%)\n",
            "  0:09:26.736   148M / 1G    INFO    General                 (path_extender.hpp         : 885)   Processed 129 paths from 420 (30%)\n",
            "  0:09:26.739   148M / 1G    INFO    General                 (path_extender.hpp         : 885)   Processed 172 paths from 420 (40%)\n",
            "  0:09:26.745   148M / 1G    INFO    General                 (path_extender.hpp         : 885)   Processed 215 paths from 420 (51%)\n",
            "  0:09:26.752   148M / 1G    INFO    General                 (path_extender.hpp         : 883)   Processed 256 paths from 420 (60%)\n",
            "  0:09:26.753   148M / 1G    INFO    General                 (path_extender.hpp         : 885)   Processed 258 paths from 420 (61%)\n",
            "  0:09:26.756   148M / 1G    INFO    General                 (path_extender.hpp         : 885)   Processed 301 paths from 420 (71%)\n",
            "  0:09:26.757   148M / 1G    INFO    General                 (path_extender.hpp         : 885)   Processed 344 paths from 420 (81%)\n",
            "  0:09:26.757   148M / 1G    INFO    General                 (path_extender.hpp         : 885)   Processed 387 paths from 420 (92%)\n",
            "  0:09:26.757   148M / 1G    INFO    General                 (launcher.cpp              : 234)   Finalizing paths\n",
            "  0:09:26.757   148M / 1G    INFO    General                 (launcher.cpp              : 236)   Deduplicating paths\n",
            "  0:09:26.759   148M / 1G    INFO    General                 (launcher.cpp              : 240)   Paths deduplicated\n",
            "  0:09:26.759   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 295)   Removing overlaps\n",
            "  0:09:26.759   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 298)   Sorting paths\n",
            "  0:09:26.759   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 305)   Marking overlaps\n",
            "  0:09:26.759   148M / 1G    INFO   OverlapRemover           (pe_resolver.hpp           : 130)   Marking start/end overlaps\n",
            "  0:09:26.761   148M / 1G    INFO   OverlapRemover           (pe_resolver.hpp           : 133)   Marking remaining overlaps\n",
            "  0:09:26.763   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 308)   Splitting paths\n",
            "  0:09:26.765   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 313)   Deduplicating paths\n",
            "  0:09:26.766   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 315)   Overlaps removed\n",
            "  0:09:26.767   148M / 1G    INFO    General                 (launcher.cpp              : 257)   Paths finalized\n",
            "  0:09:26.767   148M / 1G    INFO    General                 (launcher.cpp              : 427)   Closing gaps in paths\n",
            "  0:09:26.768   148M / 1G    INFO    General                 (launcher.cpp              : 455)   Gap closing completed\n",
            "  0:09:26.769   148M / 1G    INFO    General                 (launcher.cpp              : 286)   Traversing tandem repeats\n",
            "  0:09:26.772   148M / 1G    INFO    General                 (launcher.cpp              : 296)   Traversed 9 loops\n",
            "  0:09:26.772   148M / 1G    INFO    General                 (launcher.cpp              : 234)   Finalizing paths\n",
            "  0:09:26.772   148M / 1G    INFO    General                 (launcher.cpp              : 236)   Deduplicating paths\n",
            "  0:09:26.772   148M / 1G    INFO    General                 (launcher.cpp              : 240)   Paths deduplicated\n",
            "  0:09:26.773   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 295)   Removing overlaps\n",
            "  0:09:26.773   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 298)   Sorting paths\n",
            "  0:09:26.773   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 305)   Marking overlaps\n",
            "  0:09:26.773   148M / 1G    INFO   OverlapRemover           (pe_resolver.hpp           : 130)   Marking start/end overlaps\n",
            "  0:09:26.773   148M / 1G    INFO   OverlapRemover           (pe_resolver.hpp           : 133)   Marking remaining overlaps\n",
            "  0:09:26.774   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 308)   Splitting paths\n",
            "  0:09:26.774   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 313)   Deduplicating paths\n",
            "  0:09:26.774   148M / 1G    INFO   PEResolver               (pe_resolver.hpp           : 315)   Overlaps removed\n",
            "  0:09:26.775   148M / 1G    INFO    General                 (launcher.cpp              : 257)   Paths finalized\n",
            "  0:09:26.775   148M / 1G    INFO    General                 (launcher.cpp              : 534)   ExSPAnder repeat resolving tool finished\n",
            "  0:09:26.787   136M / 1G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
            "  0:09:26.788   136M / 1G    INFO    General                 (contig_output_stage.cpp   :  45)   Writing GFA to /content/spade_ecoli//K55/assembly_graph_with_scaffolds.gfa\n",
            "  0:09:26.857   136M / 1G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /content/spade_ecoli//K55/before_rr.fasta\n",
            "  0:09:26.949   136M / 1G    INFO    General                 (contig_output_stage.cpp   :  56)   Outputting FastG graph to /content/spade_ecoli//K55/assembly_graph.fastg\n",
            "  0:09:27.204   136M / 1G    INFO    General                 (contig_output_stage.cpp   :  20)   Outputting FastG paths to /content/spade_ecoli//K55/final_contigs.paths\n",
            "  0:09:27.322   136M / 1G    INFO    General                 (contig_output_stage.cpp   :  20)   Outputting FastG paths to /content/spade_ecoli//K55/scaffolds.paths\n",
            "  0:09:27.422   136M / 1G    INFO    General                 (launch.hpp                : 151)   SPAdes finished\n",
            "  0:09:27.513    12M / 1G    INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 9 minutes 27 seconds\n",
            "\n",
            "===== Assembling finished. Used k-mer sizes: 21, 33, 55 \n",
            "\n",
            " * Assembled contigs are in /content/spade_ecoli/contigs.fasta\n",
            " * Assembled scaffolds are in /content/spade_ecoli/scaffolds.fasta\n",
            " * Paths in the assembly graph corresponding to the contigs are in /content/spade_ecoli/contigs.paths\n",
            " * Paths in the assembly graph corresponding to the scaffolds are in /content/spade_ecoli/scaffolds.paths\n",
            " * Assembly graph is in /content/spade_ecoli/assembly_graph.fastg\n",
            " * Assembly graph in GFA format is in /content/spade_ecoli/assembly_graph_with_scaffolds.gfa\n",
            "\n",
            "======= SPAdes pipeline finished.\n",
            "\n",
            "SPAdes log can be found here: /content/spade_ecoli/spades.log\n",
            "\n",
            "Thank you for using SPAdes!\n"
          ]
        }
      ],
      "source": [
        "# Run SPAdes for E. coli genome assembly\n",
        "\n",
        "# because if you don't mkdir, bwa will say \"no such file or directory\"\n",
        "!sudo mkdir -p spade_ecoli\n",
        "\n",
        "!spades.py -s ecoli_reads.fastq -o spade_ecoli --only-assembler -k 21,33,55"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gNnSijVzu0GE"
      },
      "source": [
        "\n",
        "After assembling the genome, SPAdes will output several files in the specified directory, including the main contig file. We can assess the quality of the assembly using metrics like N50, total contig length, and the number of contigs.\n",
        "\n",
        "### Explain this Key Metrics:\n",
        "\n",
        "1. **N50**:\n",
        "\n",
        "\n",
        "-> Half of the genome sequence is covered by contigs larger than or equal the N50 contig size.\n",
        "\n",
        " It measures the contiguity of the assembly. The greatest the number, the more contiguous the assembly is. It is calculated by sorting the contigs from longest to shortest, then summing the length of the contigs until 50% of the total assembly size is reached. The N50 is the length of the contig in which this threshold is reached.\n",
        "\n",
        "2. **Total Length**:\n",
        "\n",
        "from samtools docs:\n",
        " total length - number of processed bases from reads that are neither secondary nor supplementary (flags 0x100 (256) and 0x800 (2048) are not set).\n",
        "\n",
        "Sum of all contig lengths. It shows what percentage of the genome length was assembled.\n",
        "\n",
        "3. **Number of Contigs**:\n",
        "\n",
        "A contig (from contiguous) is a set of overlapping DNA segments that together represent a consensus region of DNA.\n",
        "\n",
        "When creating a draft genome, individual reads of DNA are second assembled into contigs, which, by the nature of their assembly, have gaps between them.\n",
        "\n",
        "So number of contigs is number of gaps + 1\n",
        "\n",
        "If coverage is C, assuming the number of reads starting at position i follows the poisson distribution with lambda = C, the expected number of contigs will be N * exp(-C), where N is the number of reads.\n",
        "\n",
        "\n",
        "4. **Genome Coverage**:\n",
        "\n",
        "The average number of reads over a nucleotide in the genome\n",
        "\n",
        "C = LN / G\n",
        "\n",
        "C stands for coverage -\n",
        "G is the haploid genome length -\n",
        "L is the read length -\n",
        "N is the number of reads\n",
        "\n",
        "more coverage -> less gaps -> longer and fewer contigs\n",
        "\n",
        "<br>\n",
        "\n",
        "https://mathweb.ucsd.edu/~gptesler/283/slides/shotgun_f19-handout.pdf\n",
        "\n",
        "https://en.wikipedia.org/wiki/N50,_L50,_and_related_statistics\n",
        "\n",
        "https://help.biobam.com/space/OUM/3501034617/Genome+Assembly+Quality+Assessment\n",
        "\n",
        "https://www.researchgate.net/post/How-to-calculate-the-genome-coverage-and-duplication-ratio-of-assembly-contigs\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "e-8QVfKiXMSZ"
      },
      "source": [
        "---\n",
        "\n",
        "## 2. Download and Normalize *Citrobacter* Reads\n",
        "\n",
        "The next step involves downloading the reads for *Citrobacter*, performing adapter trimming, and quality control using `fastp`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Fs-I-8hdu3Pb",
        "outputId": "477657c3-a511-4358-a1b0-3ce01a73e753"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-11-27 20:52:03--  https://trace.ncbi.nlm.nih.gov/Traces/sra-reads-be/fastq?acc=SRR31179981\n",
            "Resolving trace.ncbi.nlm.nih.gov (trace.ncbi.nlm.nih.gov)... 130.14.29.113, 2607:f220:41e:4290::113\n",
            "Connecting to trace.ncbi.nlm.nih.gov (trace.ncbi.nlm.nih.gov)|130.14.29.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: unspecified [application/x-gzip]\n",
            "Saving to: ‘citrobacter_reads.fastq’\n",
            "\n",
            "citrobacter_reads.f     [               <=>  ] 106.91M  1.23MB/s    in 89s     \n",
            "\n",
            "2024-11-27 20:53:34 (1.20 MB/s) - ‘citrobacter_reads.fastq’ saved [112099346]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "# Download Citrobacter reads\n",
        "!wget -O citrobacter_reads.fastq https://trace.ncbi.nlm.nih.gov/Traces/sra-reads-be/fastq?acc=SRR31179981"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lUDFhVdPBWmG",
        "outputId": "9e2ee6dc-8722-4738-b499-731ddad5145f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following NEW packages will be installed:\n",
            "  fastp\n",
            "0 upgraded, 1 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 193 kB of archives.\n",
            "After this operation, 640 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 fastp amd64 0.20.1+dfsg-1 [193 kB]\n",
            "Fetched 193 kB in 1s (214 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 1.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package fastp.\n",
            "(Reading database ... 125531 files and directories currently installed.)\n",
            "Preparing to unpack .../fastp_0.20.1+dfsg-1_amd64.deb ...\n",
            "Unpacking fastp (0.20.1+dfsg-1) ...\n",
            "Setting up fastp (0.20.1+dfsg-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following NEW packages will be installed:\n",
            "  seqtk\n",
            "0 upgraded, 1 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 30.2 kB of archives.\n",
            "After this operation, 85.0 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 seqtk amd64 1.3-2 [30.2 kB]\n",
            "Fetched 30.2 kB in 0s (63.9 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 1.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package seqtk.\n",
            "(Reading database ... 125540 files and directories currently installed.)\n",
            "Preparing to unpack .../archives/seqtk_1.3-2_amd64.deb ...\n",
            "Unpacking seqtk (1.3-2) ...\n",
            "Setting up seqtk (1.3-2) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Detecting adapter sequence for read1...\n",
            "No adapter detected for read1\n",
            "\n",
            "Read1 before filtering:\n",
            "total reads: 1004342\n",
            "total bases: 302306942\n",
            "Q20 bases: 302306942(100%)\n",
            "Q30 bases: 302306942(100%)\n",
            "\n",
            "Read1 after filtering:\n",
            "total reads: 1004311\n",
            "total bases: 302295929\n",
            "Q20 bases: 302295929(100%)\n",
            "Q30 bases: 302295929(100%)\n",
            "\n",
            "Filtering result:\n",
            "reads passed filter: 1004311\n",
            "reads failed due to low quality: 0\n",
            "reads failed due to too many N: 31\n",
            "reads failed due to too short: 0\n",
            "reads with adapter trimmed: 0\n",
            "bases trimmed due to adapters: 0\n",
            "\n",
            "Duplication rate (may be overestimated since this is SE data): 10.7541%\n",
            "\n",
            "JSON report: citro.json\n",
            "HTML report: citro.html\n",
            "\n",
            "fastp -i citro_fixed.fastq -o citro_trimmed.fastq -h citro.html -j citro.json --detect_adapter_for_pe --trim_poly_g --cut_front --cut_tail --cut_right --cut_mean_quality 20 \n",
            "fastp v0.20.1, time used: 36 seconds\n"
          ]
        }
      ],
      "source": [
        "# Doing some Normalization and triming adaptors\n",
        "!sudo apt-get install fastp\n",
        "!sudo apt-get install seqtk\n",
        "\n",
        "!seqtk seq citrobacter_reads.fastq > citro_fixed.fastq\n",
        "!fastp -i citro_fixed.fastq -o citro_trimmed.fastq -h citro.html -j citro.json --detect_adapter_for_pe --trim_poly_g --cut_front --cut_tail --cut_right --cut_mean_quality 20"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3mTovzYKu-3n"
      },
      "source": [
        "The `fastp` tool allows us to remove low-quality reads and adapters that could negatively impact alignment accuracy in the following steps.\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "orpd0R5DXP6T"
      },
      "source": [
        "---\n",
        "\n",
        "## 3. Align Trimmed *Citrobacter* Reads to Assembled *E. coli* Genome\n",
        "\n",
        "Using BWA, we align the trimmed *Citrobacter* reads to the *E. coli* genome to evaluate the similarity between these two bacterial genomes."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PNY1LBhYdE80",
        "outputId": "8bfeda01-29a6-4570-eb61-c9ab02a0ac62"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting Bio\n",
            "  Downloading bio-1.7.1-py3-none-any.whl.metadata (5.7 kB)\n",
            "Collecting biopython>=1.80 (from Bio)\n",
            "  Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)\n",
            "Collecting gprofiler-official (from Bio)\n",
            "  Downloading gprofiler_official-1.0.0-py3-none-any.whl.metadata (11 kB)\n",
            "Collecting mygene (from Bio)\n",
            "  Downloading mygene-3.2.2-py2.py3-none-any.whl.metadata (10 kB)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from Bio) (2.2.2)\n",
            "Requirement already satisfied: pooch in /usr/local/lib/python3.10/dist-packages (from Bio) (1.8.2)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from Bio) (2.32.3)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from Bio) (4.66.6)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython>=1.80->Bio) (1.26.4)\n",
            "Collecting biothings-client>=0.2.6 (from mygene->Bio)\n",
            "  Downloading biothings_client-0.3.1-py2.py3-none-any.whl.metadata (9.8 kB)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas->Bio) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->Bio) (2024.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/dist-packages (from pandas->Bio) (2024.2)\n",
            "Requirement already satisfied: platformdirs>=2.5.0 in /usr/local/lib/python3.10/dist-packages (from pooch->Bio) (4.3.6)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from pooch->Bio) (24.2)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (3.4.0)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (2.2.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (2024.8.30)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas->Bio) (1.16.0)\n",
            "Downloading bio-1.7.1-py3-none-any.whl (280 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m281.0/281.0 kB\u001b[0m \u001b[31m5.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m52.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
            "Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
            "Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n",
            "Installing collected packages: biopython, gprofiler-official, biothings-client, mygene, Bio\n",
            "Successfully installed Bio-1.7.1 biopython-1.84 biothings-client-0.3.1 gprofiler-official-1.0.0 mygene-3.2.2\n"
          ]
        }
      ],
      "source": [
        "!pip install Bio"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kVUZcDBUvE2f",
        "outputId": "ff91662e-78c7-49a7-f1bd-6fa16559a3e2"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bwa_index] Pack FASTA... 0.03 sec\n",
            "[bwa_index] Construct BWT for the packed sequence...\n",
            "[bwa_index] 2.54 seconds elapse.\n",
            "[bwa_index] Update BWT... 0.03 sec\n",
            "[bwa_index] Pack forward-only FASTA... 0.02 sec\n",
            "[bwa_index] Construct SA from BWT and Occ... 0.83 sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa index spade_ecoli/contigs.fasta\n",
            "[main] Real time: 3.623 sec; CPU: 3.455 sec\n"
          ]
        }
      ],
      "source": [
        "# Index the assembled E. coli genome (assumed in contigs.fasta)\n",
        "!bwa index spade_ecoli/contigs.fasta"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "gkIfkyHjvGfa",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "b43ce3cf-c770-4849-8c42-1a78ece56400"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 33224 sequences (10000300 bp)...\n",
            "[M::process] read 33224 sequences (10000366 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 13.136 CPU sec, 20.515 real sec\n",
            "[M::process] read 33224 sequences (10000291 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 12.167 CPU sec, 12.177 real sec\n",
            "[M::process] read 33224 sequences (10000386 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 11.804 CPU sec, 11.959 real sec\n",
            "[M::process] read 33224 sequences (10000313 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 13.892 CPU sec, 13.983 real sec\n",
            "[M::process] read 33224 sequences (10000336 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 11.120 CPU sec, 11.061 real sec\n",
            "[M::process] read 33224 sequences (10000360 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 13.973 CPU sec, 14.217 real sec\n",
            "[M::process] read 33224 sequences (10000372 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 12.748 CPU sec, 13.188 real sec\n",
            "[M::process] read 33224 sequences (10000314 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 11.352 CPU sec, 11.451 real sec\n",
            "[M::process] read 33224 sequences (10000374 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 13.798 CPU sec, 13.931 real sec\n",
            "[M::process] read 33224 sequences (10000337 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 11.478 CPU sec, 11.460 real sec\n",
            "[M::process] read 33224 sequences (10000353 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 17.161 CPU sec, 17.823 real sec\n",
            "[M::process] read 33224 sequences (10000356 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 14.417 CPU sec, 14.662 real sec\n",
            "[M::process] read 33224 sequences (10000406 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 10.016 CPU sec, 9.925 real sec\n",
            "[M::process] read 33224 sequences (10000356 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 14.897 CPU sec, 15.259 real sec\n",
            "[M::process] read 33224 sequences (10000375 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 12.247 CPU sec, 12.246 real sec\n",
            "[M::process] read 33224 sequences (10000381 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 15.965 CPU sec, 16.990 real sec\n",
            "[M::process] read 33224 sequences (10000394 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 14.299 CPU sec, 14.615 real sec\n",
            "[M::process] read 33224 sequences (10000297 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 11.300 CPU sec, 11.389 real sec\n",
            "[M::process] read 33224 sequences (10000400 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 14.557 CPU sec, 14.749 real sec\n",
            "[M::process] read 33224 sequences (10000371 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 11.883 CPU sec, 11.896 real sec\n",
            "[M::process] read 33224 sequences (10000411 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 13.816 CPU sec, 14.067 real sec\n",
            "[M::process] read 33224 sequences (10000368 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 12.703 CPU sec, 12.742 real sec\n",
            "[M::process] read 33224 sequences (10000403 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 12.385 CPU sec, 12.640 real sec\n",
            "[M::process] read 33224 sequences (10000404 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 14.430 CPU sec, 14.603 real sec\n",
            "[M::process] read 33224 sequences (10000391 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 11.801 CPU sec, 11.891 real sec\n",
            "[M::process] read 33224 sequences (10000414 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 14.938 CPU sec, 15.149 real sec\n",
            "[M::process] read 33224 sequences (10000391 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 11.544 CPU sec, 11.518 real sec\n",
            "[M::process] read 33224 sequences (10000413 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 14.328 CPU sec, 14.502 real sec\n",
            "[M::process] read 33224 sequences (10000405 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 12.631 CPU sec, 12.697 real sec\n",
            "[M::process] read 7591 sequences (2284891 bp)...\n",
            "[M::mem_process_seqs] Processed 33224 reads in 12.991 CPU sec, 13.198 real sec\n",
            "[M::mem_process_seqs] Processed 7591 reads in 2.379 CPU sec, 2.299 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem spade_ecoli/contigs.fasta citro_trimmed.fastq\n",
            "[main] Real time: 409.119 sec; CPU: 396.276 sec\n"
          ]
        }
      ],
      "source": [
        "# Align trimmed Citrobacter reads to the E. coli assembled genome\n",
        "!bwa mem spade_ecoli/contigs.fasta citro_trimmed.fastq > citro_ecoli.sam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2ClpJaB2vR8t",
        "outputId": "313a5c80-c090-4b19-8c52-1dd6e01749c2"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "samtools is already the newest version (1.13-4).\n",
            "samtools set to manually installed.\n",
            "0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.\n"
          ]
        }
      ],
      "source": [
        "# Install samtools if not already installed\n",
        "!apt-get install samtools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "k4A01mqivJ1-"
      },
      "outputs": [],
      "source": [
        "# Convert SAM to BAM\n",
        "!samtools view -S -b citro_ecoli.sam > citro_ecoli.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "0gskwcBKvKno"
      },
      "outputs": [],
      "source": [
        "# Sort the BAM file\n",
        "!samtools sort citro_ecoli.bam -o sorted_citro_ecoli.bam"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gR4RMPLOvMgQ"
      },
      "source": [
        "---\n",
        "\n",
        "## 4. Calculate Similarity Using Alignment Statistics\n",
        "\n",
        "The alignment statistics provide insights into the similarity between *E. coli* and *Citrobacter*.\n",
        "\n",
        "Explain key metrics from the `samtools` flagstat output:\n",
        "<br>\n",
        "<br>\n",
        "\n",
        "- **Total Reads Mapped**:\n",
        " reads mapped - number of reads, paired or single, that are mapped (flag 0x4 or 0x8 not set).\n",
        " That includes multimappers, so it won't necessarily be the number of mapped reads.\n",
        "\n",
        "command format:\n",
        "\n",
        " samtools flagstat maps/bt2.global.sam\n",
        "\n",
        " <br>\n",
        "\n",
        "- **Coverage Depth**:\n",
        "\n",
        "number of times each nucleotide position appeared in reads. Can be obtained by these commands:\n",
        "\n",
        "- get coverage of whole alignment file\n",
        "\n",
        "samtools coverage bt2.g.sorted.bam\n",
        "\n",
        "- get coverage of a particular sequence\n",
        "\n",
        "samtools coverage bt2.g.sorted.bam -r NC_003283.11\n",
        "\n",
        "- get coverage of a particular sequence given initial and end position\n",
        "\n",
        "samtools coverage bt2.g.sorted.bam -r NC_003280.10:1000-25000\n",
        "\n",
        "- get coverage as a histogram\n",
        "\n",
        "samtools coverage bt2.g.sorted.bam --histogram\n",
        "\n",
        "- get coverage as depth plot\n",
        "\n",
        "samtools coverage bt2.g.sorted.bam --plot-depth\n",
        "\n",
        "- store coverage in a tsv/tab-separated file to open as Excel file\n",
        "\n",
        "samtools coverage bt2.g.sorted.bam > coverage.tsv\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "https://www.biostars.org/p/138116/\n",
        "\n",
        "http://www.htslib.org/doc/samtools-stats.html\n",
        "\n",
        "https://medium.com/@manabeel.vet/a-beginners-guide-to-genomic-data-analysis-samtools-and-genome-mapping-visualization-in-d2f81ed907a7\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "Sjm0FohyvUcE"
      },
      "outputs": [],
      "source": [
        "# Generate alignment statistics\n",
        "!samtools flagstat sorted_citro_ecoli.bam > align_stats.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "id": "_S9ZA7Z4vYX7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d2276346-a487-4170-d07e-a0aafbfd8407"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1023752 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "1004311 + 0 primary\n",
            "0 + 0 secondary\n",
            "19441 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "485726 + 0 mapped (47.45% : N/A)\n",
            "466285 + 0 primary mapped (46.43% : N/A)\n",
            "0 + 0 paired in sequencing\n",
            "0 + 0 read1\n",
            "0 + 0 read2\n",
            "0 + 0 properly paired (N/A : N/A)\n",
            "0 + 0 with itself and mate mapped\n",
            "0 + 0 singletons (N/A : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ],
      "source": [
        "# Show alignment statistics file\n",
        "!cat align_stats.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "id": "1_I_JdeVwC7D",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "74ca9a89-fe7d-4aed-b110-d2f6109e9d13"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'total_reads': 1023752,\n",
              " 'mapped_reads': 466285,\n",
              " 'similarity_percentage': 45.54667536669037}"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ],
      "source": [
        "import re\n",
        "from math import inf\n",
        "\n",
        "# Function to calculate similarity based on alignment statistics\n",
        "def calculate_similarity(alignment_stats_file):\n",
        "\n",
        "  with open(alignment_stats_file, 'r') as f:\n",
        "    lines = f.readlines()\n",
        "\n",
        "  total_reads = 0\n",
        "  mapped_reads = 0\n",
        "\n",
        "  ## WRITE YOUR CODE HERE\n",
        "  similarity_percentage = -inf\n",
        "\n",
        "  total_pat = r'(\\d+)(?:\\s+\\+\\s+0)?\\s+in total'\n",
        "  map_pat = r'(\\d+)(?:\\s+\\+\\s+0)?\\s+primary mapped '\n",
        "\n",
        "  for line in lines:\n",
        "    total_match = re.search(total_pat, line)\n",
        "    if total_match != None:\n",
        "        total_reads = int(total_match.groups(0)[0])\n",
        "        # print(total_reads)\n",
        "\n",
        "    map_match = re.search(map_pat, line)\n",
        "    if map_match != None:\n",
        "        mapped_reads = int(map_match.groups(0)[0])\n",
        "        # print(mapped_reads)\n",
        "\n",
        "    if total_reads > 0:\n",
        "        similarity_percentage = mapped_reads * 100 / total_reads\n",
        "  ##\n",
        "\n",
        "  return {\n",
        "      \"total_reads\": total_reads,\n",
        "      \"mapped_reads\": mapped_reads,\n",
        "      \"similarity_percentage\": similarity_percentage\n",
        "  }\n",
        "\n",
        "\n",
        "similarity_metrics = calculate_similarity(\"align_stats.txt\")\n",
        "similarity_metrics"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "h_2LQWFZSmBX"
      },
      "source": [
        "## Additional Topics\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lVgskuvLXv-N"
      },
      "source": [
        "\n",
        "### Assembly Challenges\n",
        "\n",
        "1. Genome assembly for bacterial genomes faces challenges such as **Repetitive Sequences, Short Reads & Chimeric Reads** , explain them:\n",
        "\n",
        "Repetitive DNA sequences are segments of the genome where specific patterns of nucleotides occur multiple times in succession.\n",
        "\n",
        "They may result in fragmented assemblies or misassembled contigs because assembly algorithms may not determinstically find the order of their occurences soley based on read data.\n",
        "to solve this problem, use long-read sequencing technologies (e.g., PacBio or Oxford Nanopore) that generate reads long enough to span repeats.\n",
        "<br>\n",
        "\n",
        "The limitations of short-read sequencing:\n",
        "\n",
        " As they are short, it may not be possible to map reads to the specific region of the reference genome they came from. If reads cannot be mapped to the reference genome, they may be discarded, leading to a gap in sequencing data.\n",
        "\n",
        "- Genes with a pseudogene or genes in repetitive regions can therefore be difficult to sequence.\n",
        "- Repeat expansion disorders can be difficult to detect accurately, especially if the length of the short tandem repeat region exceeds the read length. This is why these are usually tested for separately.\n",
        "- In addition, large copy number variants and structural variants are not usually detected by short-read sequencing as effectively as they are detected by gold-standard methods such as arrays or karyotyping.\n",
        "- Methylation changes are not detected, which require methylation studies.\n",
        "\n",
        "Long-read sequencing is gaining ground in clinical applications as it overcomes some of these issues.\n",
        "<br>\n",
        "\n",
        "Chimeric reads:\n",
        "\n",
        "A chimera in DNA sequencing is basically when your polymerase has synthesized a new strand of DNA from two different parent strands of DNA during your PCR. So the resulting DNA sequence is not representative of any parent DNA sequences. This is why you screen them out, especially for ecology studies, as chimeras will misrepresent what (bacteria, virus, fungi, etc.) is in your sample.\n",
        "\n",
        "\n",
        "https://pmc.ncbi.nlm.nih.gov/articles/PMC7736650/\n",
        "\n",
        "https://www.genomicseducation.hee.nhs.uk/genotes/knowledge-hub/short-read-sequencing/\n",
        "\n",
        "https://www.reddit.com/r/bioinformatics/comments/ax8edr/what_is_a_chimeric_read_or_a_chimeric_alignment/\n",
        "\n",
        "<br>\n",
        "2. What other challenges bacterial genome have for assembly?\n",
        "\n",
        "- Some sequencing platforms, like Illumina, have a bias towards GC balanced regions, which can lead to uneven sequencing depth.\n",
        "- Assembling bacterial genomes can be computationally intensive.\n",
        "- There's a trade-off between assembling as many regions as possible, or fewer but more accurate sequences.\n",
        "\n",
        "Third-generation sequencing (TGS) has made it possible to assemble bacterial genomes more quickly, but most current TGS assemblers are designed for species with non-circular genomes\n",
        "\n",
        "https://pmc.ncbi.nlm.nih.gov/articles/PMC9473318/\n",
        "\n",
        "https://genomics.umn.edu/services/long-read-technologies\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "huAN9S5xXzJX"
      },
      "source": [
        "### Key Assembly Metrics\n",
        "- **Distribution of Contig Lengths**: Plotting the distribution of contig lengths allows us to visually assess assembly fragmentation."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "id": "hR2pZK5FUVGj",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 469
        },
        "outputId": "ca31eb72-b95d-4db3-c330-ed6eca746e03"
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkgAAAGwCAYAAABSN5pGAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAy6UlEQVR4nO3deXhUVZ7/8U+FkIJIFkLI1gTCjixhbdJpkaUTJYFBBboViCMojRsiEtfYtoj6dGhRBrURpkchOoooI6IjgrIjGlCWiChGiMFgk4CCpEiQmOX8/uBHjXUTIBSVVCq8X89zn4d7z6lT35Mbzee599wqmzHGCAAAAE5+3i4AAACgoSEgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAwt/bBTQEVVVVOnTokIKCgmSz2bxdDgAAqAVjjE6cOKGYmBj5+Xn2mg8BSdKhQ4cUGxvr7TIAAIAbDh48qDZt2nh0TAKSpKCgIEmnf8DBwcFergYAANSGw+FQbGys8++4JxGQJOdtteDgYAISAAA+pi6Wx7BIGwAAwIKABAAAYEFAAgAAsCAgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAgoAEAABgQUACAACwICABAABYEJAAAAAsCEgAAAAWBCQAAAALf28X0NjFPbTS2yVcsAOzR3q7BAAAvIorSAAAABYEJAAAAAsCEgAAgIVXA9LmzZs1atQoxcTEyGazacWKFS7tNputxm3OnDnOPnFxcdXaZ8+eXc8zAQAAjYlXA1Jpaal69+6t+fPn19heWFjosi1atEg2m01jx4516ff444+79Js2bVp9lA8AABoprz7FlpqaqtTU1LO2R0VFuey/8847GjZsmDp06OByPCgoqFpfAAAAd/nMGqTDhw9r5cqVmjx5crW22bNnq1WrVurbt6/mzJmjioqKc45VVlYmh8PhsgEAAJzhM5+D9PLLLysoKEhjxoxxOX733XerX79+CgsL0yeffKKMjAwVFhZq7ty5Zx0rMzNTs2bNquuSAQCAj/KZgLRo0SKlpaWpWbNmLsfT09Od/46Pj1dAQIBuu+02ZWZmym631zhWRkaGy+scDodiY2PrpnAAAOBzfCIgffTRR8rNzdUbb7xx3r4JCQmqqKjQgQMH1LVr1xr72O32s4YnAAAAn1iD9NJLL6l///7q3bv3efvm5OTIz89PERER9VAZAABojLx6BamkpET79+937ufn5ysnJ0dhYWFq27atpNO3v5YtW6Znnnmm2uuzs7O1bds2DRs2TEFBQcrOztaMGTN04403qmXLlvU2DwAA0Lh4NSBt375dw4YNc+6fWRc0ceJEZWVlSZKWLl0qY4zGjx9f7fV2u11Lly7VY489prKyMrVv314zZsxwWV8EAABwoWzGGOPtIrzN4XAoJCRExcXFCg4O9ujYcQ+t9Oh49eHA7JHeLgEAgPOqy7/fPrEGCQAAoD4RkAAAACwISAAAABYEJAAAAAsCEgAAgAUBCQAAwIKABAAAYEFAAgAAsCAgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAgoAEAABgQUACAACwICABAABYEJAAAAAsCEgAAAAWBCQAAAALAhIAAIAFAQkAAMCCgAQAAGBBQAIAALAgIAEAAFgQkAAAACwISAAAABYEJAAAAAsCEgAAgAUBCQAAwIKABAAAYEFAAgAAsCAgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAwqsBafPmzRo1apRiYmJks9m0YsUKl/ZJkybJZrO5bCkpKS59jh07prS0NAUHBys0NFSTJ09WSUlJPc4CAAA0Nl4NSKWlperdu7fmz59/1j4pKSkqLCx0bq+//rpLe1pamr788kutWbNG7733njZv3qxbb721rksHAACNmL833zw1NVWpqann7GO32xUVFVVj2969e7V69Wp99tlnGjBggCTp+eef14gRI/T0008rJibG4zUDAIDGr8GvQdq4caMiIiLUtWtX3XHHHTp69KizLTs7W6Ghoc5wJEnJycny8/PTtm3bzjpmWVmZHA6HywYAAHBGgw5IKSkpeuWVV7Ru3Tr9/e9/16ZNm5SamqrKykpJUlFRkSIiIlxe4+/vr7CwMBUVFZ113MzMTIWEhDi32NjYOp0HAADwLV69xXY+48aNc/67V69eio+PV8eOHbVx40YlJSW5PW5GRobS09Od+w6Hg5AEAACcGvQVJKsOHTooPDxc+/fvlyRFRUXpyJEjLn0qKip07Nixs65bkk6vawoODnbZAAAAzvCpgPT999/r6NGjio6OliQlJibq+PHj2rFjh7PP+vXrVVVVpYSEBG+VCQAAfJxXb7GVlJQ4rwZJUn5+vnJychQWFqawsDDNmjVLY8eOVVRUlPLy8vTAAw+oU6dOGj58uCTp8ssvV0pKiqZMmaKFCxeqvLxcd911l8aNG8cTbAAAwG1evYK0fft29e3bV3379pUkpaenq2/fvnr00UfVpEkT7d69W9dcc426dOmiyZMnq3///vroo49kt9udY7z22mvq1q2bkpKSNGLECA0aNEj//Oc/vTUlAADQCHj1CtLQoUNljDlr+wcffHDeMcLCwrRkyRJPlgUAAC5xPrUGCQAAoD4QkAAAACwISAAAABYEJAAAAAsCEgAAgAUBCQAAwIKABAAAYEFAAgAAsCAgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAgoAEAABgQUACAACwICABAABYEJAAAAAsCEgAAAAWBCQAAAALAhIAAIAFAQkAAMCCgAQAAGBBQAIAALAgIAEAAFgQkAAAACwISAAAABYEJAAAAAsCEgAAgAUBCQAAwIKABAAAYEFAAgAAsCAgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAwqsBafPmzRo1apRiYmJks9m0YsUKZ1t5ebkefPBB9erVS5dddpliYmJ000036dChQy5jxMXFyWazuWyzZ8+u55kAAIDGxKsBqbS0VL1799b8+fOrtZ08eVI7d+7UX//6V+3cuVPLly9Xbm6urrnmmmp9H3/8cRUWFjq3adOm1Uf5AACgkfL35punpqYqNTW1xraQkBCtWbPG5dg//vEPDRw4UAUFBWrbtq3zeFBQkKKiouq0VgAAcOnwqTVIxcXFstlsCg0NdTk+e/ZstWrVSn379tWcOXNUUVFxznHKysrkcDhcNgAAgDO8egXpQpw6dUoPPvigxo8fr+DgYOfxu+++W/369VNYWJg++eQTZWRkqLCwUHPnzj3rWJmZmZo1a1Z9lA0AAHyQTwSk8vJyXX/99TLGaMGCBS5t6enpzn/Hx8crICBAt912mzIzM2W322scLyMjw+V1DodDsbGxdVM8AADwOQ0+IJ0JR999953Wr1/vcvWoJgkJCaqoqNCBAwfUtWvXGvvY7fazhicAAIAGHZDOhKN9+/Zpw4YNatWq1Xlfk5OTIz8/P0VERNRDhQAAoDHyakAqKSnR/v37nfv5+fnKyclRWFiYoqOj9cc//lE7d+7Ue++9p8rKShUVFUmSwsLCFBAQoOzsbG3btk3Dhg1TUFCQsrOzNWPGDN14441q2bKlt6YFAAB8nFcD0vbt2zVs2DDn/pl1QRMnTtRjjz2md999V5LUp08fl9dt2LBBQ4cOld1u19KlS/XYY4+prKxM7du314wZM1zWFwEAAFworwakoUOHyhhz1vZztUlSv379tHXrVk+XBQAALnE+9TlIAAAA9YGABAAAYEFAAgAAsCAgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAgoAEAABgQUACAACwICABAABYEJAAAAAsCEgAAAAWBCQAAAALAhIAAIAFAQkAAMCCgAQAAGBBQAIAALAgIAEAAFgQkAAAACwISAAAABZuBaRvv/3W03UAAAA0GG4FpE6dOmnYsGF69dVXderUKU/XBAAA4FVuBaSdO3cqPj5e6enpioqK0m233aZPP/3U07UBAAB4hVsBqU+fPnr22Wd16NAhLVq0SIWFhRo0aJB69uypuXPn6ocffvB0nQAAAPXmohZp+/v7a8yYMVq2bJn+/ve/a//+/brvvvsUGxurm266SYWFhZ6qEwAAoN5cVEDavn277rzzTkVHR2vu3Lm67777lJeXpzVr1ujQoUO69tprPVUnAABAvfF350Vz587V4sWLlZubqxEjRuiVV17RiBEj5Od3Om+1b99eWVlZiouL82StAAAA9cKtgLRgwQLdcsstmjRpkqKjo2vsExERoZdeeumiigMAAPAGtwLSvn37ztsnICBAEydOdGd4AAAAr3JrDdLixYu1bNmyaseXLVuml19++aKLAgAA8Ca3AlJmZqbCw8OrHY+IiNDf/va3iy4KAADAm9wKSAUFBWrfvn214+3atVNBQcFFFwUAAOBNbgWkiIgI7d69u9rxzz//XK1atbroogAAALzJrYA0fvx43X333dqwYYMqKytVWVmp9evXa/r06Ro3bpynawQAAKhXbj3F9sQTT+jAgQNKSkqSv//pIaqqqnTTTTexBgkAAPg8twJSQECA3njjDT3xxBP6/PPP1bx5c/Xq1Uvt2rXzdH0AAAD17qK+aqRLly7605/+pH/7t39zKxxt3rxZo0aNUkxMjGw2m1asWOHSbozRo48+qujoaDVv3lzJycnVPoPp2LFjSktLU3BwsEJDQzV58mSVlJRczLQAAMAlzq0rSJWVlcrKytK6det05MgRVVVVubSvX7++VuOUlpaqd+/euuWWWzRmzJhq7U899ZSee+45vfzyy2rfvr3++te/avjw4frqq6/UrFkzSVJaWpoKCwu1Zs0alZeX6+abb9att96qJUuWuDM1AAAA9wLS9OnTlZWVpZEjR6pnz56y2WxuvXlqaqpSU1NrbDPGaN68eXrkkUecX3r7yiuvKDIyUitWrNC4ceO0d+9erV69Wp999pkGDBggSXr++ec1YsQIPf3004qJiXGrLgAAcGlzKyAtXbpUb775pkaMGOHpepzy8/NVVFSk5ORk57GQkBAlJCQoOztb48aNU3Z2tkJDQ53hSJKSk5Pl5+enbdu2afTo0TWOXVZWprKyMue+w+Gos3kAAADf49YapICAAHXq1MnTtbgoKiqSJEVGRrocj4yMdLYVFRUpIiLCpd3f319hYWHOPjXJzMxUSEiIc4uNjfVw9QAAwJe5FZDuvfdePfvsszLGeLqeepGRkaHi4mLndvDgQW+XBAAAGhC3brFt2bJFGzZs0KpVq9SjRw81bdrUpX358uUXXVhUVJQk6fDhw4qOjnYeP3z4sPr06ePsc+TIEZfXVVRU6NixY87X18Rut8tut190jQAAoHFy6wpSaGioRo8erSFDhig8PNzldlVISIhHCmvfvr2ioqK0bt065zGHw6Ft27YpMTFRkpSYmKjjx49rx44dzj7r169XVVWVEhISPFIHAAC49Lh1BWnx4sUeefOSkhLt37/fuZ+fn6+cnByFhYWpbdu2uueee/Tkk0+qc+fOzsf8Y2JidN1110mSLr/8cqWkpGjKlClauHChysvLddddd2ncuHE8wQYAANzmVkCSTt/K2rhxo/Ly8jRhwgQFBQXp0KFDCg4OVosWLWo1xvbt2zVs2DDnfnp6uiRp4sSJysrK0gMPPKDS0lLdeuutOn78uAYNGqTVq1c7PwNJkl577TXdddddSkpKkp+fn8aOHavnnnvO3WkBAADIZtxYaf3dd98pJSVFBQUFKisr0zfffKMOHTpo+vTpKisr08KFC+ui1jrjcDgUEhKi4uJiBQcHe3TsuIdWenS8+nBg9khvlwAAwHnV5d9vt9YgTZ8+XQMGDNBPP/2k5s2bO4+PHj3aZc0QAACAL3LrFttHH32kTz75RAEBAS7H4+Li9K9//csjhQEAAHiLW1eQqqqqVFlZWe34999/r6CgoIsuCgAAwJvcCkhXX3215s2b59y32WwqKSnRzJkz6/TrRwAAAOqDW7fYnnnmGQ0fPlzdu3fXqVOnNGHCBO3bt0/h4eF6/fXXPV0jAABAvXIrILVp00aff/65li5dqt27d6ukpESTJ09WWlqay6JtAAAAX+T25yD5+/vrxhtv9GQtAAAADYJbAemVV145Z/tNN93kVjEAAAANgVsBafr06S775eXlOnnypAICAhQYGEhAAgAAPs2tp9h++uknl62kpES5ubkaNGgQi7QBAIDPcysg1aRz586aPXt2tatLAAAAvsZjAUk6vXD70KFDnhwSAACg3rm1Bundd9912TfGqLCwUP/4xz90xRVXeKQwAAAAb3ErIF133XUu+zabTa1bt9Yf/vAHPfPMM56oCwAAwGvcCkhVVVWergMAAKDB8OgaJAAAgMbArStI6enpte47d+5cd94CAADAa9wKSLt27dKuXbtUXl6url27SpK++eYbNWnSRP369XP2s9lsnqkSAACgHrkVkEaNGqWgoCC9/PLLatmypaTTHx55880368orr9S9997r0SIBAADqk1trkJ555hllZmY6w5EktWzZUk8++SRPsQEAAJ/nVkByOBz64Ycfqh3/4YcfdOLEiYsuCgAAwJvcCkijR4/WzTffrOXLl+v777/X999/r7feekuTJ0/WmDFjPF0jAABAvXJrDdLChQt13333acKECSovLz89kL+/Jk+erDlz5ni0QAAAgPrmVkAKDAzUCy+8oDlz5igvL0+S1LFjR1122WUeLQ4AAMAbLuqDIgsLC1VYWKjOnTvrsssukzHGU3UBAAB4jVsB6ejRo0pKSlKXLl00YsQIFRYWSpImT57MI/4AAMDnuRWQZsyYoaZNm6qgoECBgYHO4zfccINWr17tseIAAAC8wa01SB9++KE++OADtWnTxuV4586d9d1333mkMAAAAG9x6wpSaWmpy5WjM44dOya73X7RRQEAAHiTWwHpyiuv1CuvvOLct9lsqqqq0lNPPaVhw4Z5rDgAAABvcOsW21NPPaWkpCRt375dv/zyix544AF9+eWXOnbsmD7++GNP1wgAAFCv3LqC1LNnT33zzTcaNGiQrr32WpWWlmrMmDHatWuXOnbs6OkaAQAA6tUFX0EqLy9XSkqKFi5cqL/85S91URMAAIBXXfAVpKZNm2r37t11UQsAAECD4NYtthtvvFEvvfSSp2sBAABoENxapF1RUaFFixZp7dq16t+/f7XvYJs7d65HigMAAPCGCwpI3377reLi4rRnzx7169dPkvTNN9+49LHZbJ6rDgAAwAsuKCB17txZhYWF2rBhg6TTXy3y3HPPKTIysk6KAwAA8IYLWoNkjHHZX7VqlUpLSz1aEAAAgLe5tUj7DGtgqgtxcXGy2WzVtqlTp0qShg4dWq3t9ttvr/O6AABA43VBt9jOBBDrsbr02WefqbKy0rm/Z88eXXXVVfrTn/7kPDZlyhQ9/vjjzv2avicOAACgti4oIBljNGnSJOcX0p46dUq33357tafYli9f7rECW7du7bI/e/ZsdezYUUOGDHEeCwwMVFRUVK3HLCsrU1lZmXPf4XBcfKEAAKDRuKBbbBMnTlRERIRCQkIUEhKiG2+8UTExMc79M1td+eWXX/Tqq6/qlltucbly9dprryk8PFw9e/ZURkaGTp48ec5xMjMzXeqNjY2ts5oBAIDvsZn6WEjkIW+++aYmTJiggoICxcTESJL++c9/ql27doqJidHu3bv14IMPauDAgee8ilXTFaTY2FgVFxcrODjYozXHPbTSo+PVhwOzR3q7BAAAzsvhcCgkJKRO/n679UGR3vLSSy8pNTXVGY4k6dZbb3X+u1evXoqOjlZSUpLy8vLO+sW5drvdeZsQAADA6qKeYqtP3333ndauXas///nP5+yXkJAgSdq/f399lAUAABohnwlIixcvVkREhEaOPPftn5ycHElSdHR0PVQFAAAaI5+4xVZVVaXFixdr4sSJ8vf/v5Lz8vK0ZMkSjRgxQq1atdLu3bs1Y8YMDR48WPHx8V6sGAAA+DKfCEhr165VQUGBbrnlFpfjAQEBWrt2rebNm6fS0lLFxsZq7NixeuSRR7xUKQAAaAx8IiBdffXVNX5qd2xsrDZt2uSFigAAQGPmM2uQAAAA6gsBCQAAwIKABAAAYEFAAgAAsCAgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAgoAEAABgQUACAACwICABAABYEJAAAAAsCEgAAAAWBCQAAAALAhIAAIAFAQkAAMCCgAQAAGBBQAIAALAgIAEAAFgQkAAAACwISAAAABYEJAAAAAsCEgAAgAUBCQAAwIKABAAAYEFAAgAAsCAgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAgoAEAABgQUACAACwICABAABYEJAAAAAsGnRAeuyxx2Sz2Vy2bt26OdtPnTqlqVOnqlWrVmrRooXGjh2rw4cPe7FiAADQGDTogCRJPXr0UGFhoXPbsmWLs23GjBn63//9Xy1btkybNm3SoUOHNGbMGC9WCwAAGgN/bxdwPv7+/oqKiqp2vLi4WC+99JKWLFmiP/zhD5KkxYsX6/LLL9fWrVv1u9/97qxjlpWVqayszLnvcDg8XzgAAPBZDf4K0r59+xQTE6MOHTooLS1NBQUFkqQdO3aovLxcycnJzr7dunVT27ZtlZ2dfc4xMzMzFRIS4txiY2PrdA4AAMC3NOiAlJCQoKysLK1evVoLFixQfn6+rrzySp04cUJFRUUKCAhQaGioy2siIyNVVFR0znEzMjJUXFzs3A4ePFiHswAAAL6mQd9iS01Ndf47Pj5eCQkJateund588001b97c7XHtdrvsdrsnSgQAAI1Qg76CZBUaGqouXbpo//79ioqK0i+//KLjx4+79Dl8+HCNa5YAAABqy6cCUklJifLy8hQdHa3+/furadOmWrdunbM9NzdXBQUFSkxM9GKVAADA1zXoW2z33XefRo0apXbt2unQoUOaOXOmmjRpovHjxyskJESTJ09Wenq6wsLCFBwcrGnTpikxMfGcT7ABAACcT4MOSN9//73Gjx+vo0ePqnXr1ho0aJC2bt2q1q1bS5L+4z/+Q35+fho7dqzKyso0fPhwvfDCC16uGgAA+DqbMcZ4uwhvczgcCgkJUXFxsYKDgz06dtxDKz06Xn04MHukt0sAAOC86vLvt0+tQQIAAKgPBCQAAAALAhIAAIAFAQkAAMCCgAQAAGBBQAIAALAgIAEAAFgQkAAAACwISAAAABYEJAAAAAsCEgAAgAUBCQAAwIKABAAAYEFAAgAAsCAgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAgoAEAABgQUACAACwICABAABYEJAAAAAsCEgAAAAWBCQAAAALAhIAAIAFAQkAAMCCgAQAAGBBQAIAALAgIAEAAFgQkAAAACwISAAAABYEJAAAAAsCEgAAgAUBCQAAwIKABAAAYEFAAgAAsGjQASkzM1O//e1vFRQUpIiICF133XXKzc116TN06FDZbDaX7fbbb/dSxQAAoDFo0AFp06ZNmjp1qrZu3ao1a9aovLxcV199tUpLS136TZkyRYWFhc7tqaee8lLFAACgMfD3dgHnsnr1apf9rKwsRUREaMeOHRo8eLDzeGBgoKKiouq7PAAA0Eg16CtIVsXFxZKksLAwl+OvvfaawsPD1bNnT2VkZOjkyZPnHKesrEwOh8NlAwAAOKNBX0H6taqqKt1zzz264oor1LNnT+fxCRMmqF27doqJidHu3bv14IMPKjc3V8uXLz/rWJmZmZo1a1Z9lA0AAHyQzRhjvF1Ebdxxxx1atWqVtmzZojZt2py13/r165WUlKT9+/erY8eONfYpKytTWVmZc9/hcCg2NlbFxcUKDg72aN1xD6306Hj14cDskd4uAQCA83I4HAoJCamTv98+cQXprrvu0nvvvafNmzefMxxJUkJCgiSdMyDZ7XbZ7XaP1wkAABqHBh2QjDGaNm2a3n77bW3cuFHt27c/72tycnIkSdHR0XVcHQAAaKwadECaOnWqlixZonfeeUdBQUEqKiqSJIWEhKh58+bKy8vTkiVLNGLECLVq1Uq7d+/WjBkzNHjwYMXHx3u5egAA4KsadEBasGCBpNMfBvlrixcv1qRJkxQQEKC1a9dq3rx5Ki0tVWxsrMaOHatHHnnEC9UCAIDGokEHpPOtH4+NjdWmTZvqqRoAAHCp8KnPQQIAAKgPBCQAAAALAhIAAIAFAQkAAMCCgAQAAGBBQAIAALAgIAEAAFgQkAAAACwISAAAABYEJAAAAAsCEgAAgAUBCQAAwIKABAAAYEFAAgAAsPD3dgFoeOIeWuntEi7YgdkjvV0CAKAR4QoSAACABQEJAADAgoAEAABgQUACAACwICABAABY8BQbGgWevAMAeBJXkAAAACwISAAAABYEJAAAAAsCEgAAgAUBCQAAwIKABAAAYEFAAgAAsCAgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAgoAEAABgQUACAACw8Pd2AcClKu6hld4uAQ3YgdkjvV0CcEnjChIAAIBFowlI8+fPV1xcnJo1a6aEhAR9+umn3i4JAAD4qEYRkN544w2lp6dr5syZ2rlzp3r37q3hw4fryJEj3i4NAAD4oEaxBmnu3LmaMmWKbr75ZknSwoULtXLlSi1atEgPPfSQl6sDAMBzfHH9oi+uqfP5gPTLL79ox44dysjIcB7z8/NTcnKysrOza3xNWVmZysrKnPvFxcWSJIfD4fH6qspOenxMAI1fXfz/CI2DL/5dqavf5zPjGmM8PrbPB6Qff/xRlZWVioyMdDkeGRmpr7/+usbXZGZmatasWdWOx8bG1kmNAHChQuZ5uwLAc+r69/nEiRMKCQnx6Jg+H5DckZGRofT0dOd+VVWVjh07platWslms3nsfRwOh2JjY3Xw4EEFBwd7bNyG5lKY56UwR+nSmOelMEfp0pgnc2w83J2nMUYnTpxQTEyMx2vy+YAUHh6uJk2a6PDhwy7HDx8+rKioqBpfY7fbZbfbXY6FhobWVYkKDg5u1L/YZ1wK87wU5ihdGvO8FOYoXRrzZI6Nhzvz9PSVozN8/im2gIAA9e/fX+vWrXMeq6qq0rp165SYmOjFygAAgK/y+StIkpSenq6JEydqwIABGjhwoObNm6fS0lLnU20AAAAXolEEpBtuuEE//PCDHn30URUVFalPnz5avXp1tYXb9c1ut2vmzJnVbuc1NpfCPC+FOUqXxjwvhTlKl8Y8mWPj0RDnaTN18WwcAACAD/P5NUgAAACeRkACAACwICABAABYEJAAAAAsCEh1aP78+YqLi1OzZs2UkJCgTz/91NslSZIee+wx2Ww2l61bt27O9lOnTmnq1Klq1aqVWrRoobFjx1b7IM6CggKNHDlSgYGBioiI0P3336+KigqXPhs3blS/fv1kt9vVqVMnZWVlVavFUz+jzZs3a9SoUYqJiZHNZtOKFStc2o0xevTRRxUdHa3mzZsrOTlZ+/btc+lz7NgxpaWlKTg4WKGhoZo8ebJKSkpc+uzevVtXXnmlmjVrptjYWD311FPValm2bJm6deumZs2aqVevXnr//fcvuBZ35zlp0qRq5zYlJcWn5pmZmanf/va3CgoKUkREhK677jrl5ua69GlIv6O1qcWdOQ4dOrTaubz99tt9Zo6StGDBAsXHxzs//C8xMVGrVq26oHF9fY6N4TxazZ49WzabTffcc88Fje1r85RBnVi6dKkJCAgwixYtMl9++aWZMmWKCQ0NNYcPH/Z2aWbmzJmmR48eprCw0Ln98MMPzvbbb7/dxMbGmnXr1pnt27eb3/3ud+b3v/+9s72iosL07NnTJCcnm127dpn333/fhIeHm4yMDGefb7/91gQGBpr09HTz1Vdfmeeff940adLErF692tnHkz+j999/3/zlL38xy5cvN5LM22+/7dI+e/ZsExISYlasWGE+//xzc80115j27dubn3/+2dknJSXF9O7d22zdutV89NFHplOnTmb8+PHO9uLiYhMZGWnS0tLMnj17zOuvv26aN29u/vM//9PZ5+OPPzZNmjQxTz31lPnqq6/MI488Ypo2bWq++OKLC6rF3XlOnDjRpKSkuJzbY8eOufRp6PMcPny4Wbx4sdmzZ4/JyckxI0aMMG3btjUlJSXOPg3pd/R8tbg7xyFDhpgpU6a4nMvi4mKfmaMxxrz77rtm5cqV5ptvvjG5ubnm4YcfNk2bNjV79uxpFOexNnNsDOfx1z799FMTFxdn4uPjzfTp02s9tq/N0xhjCEh1ZODAgWbq1KnO/crKShMTE2MyMzO9WNVpM2fONL17966x7fjx46Zp06Zm2bJlzmN79+41kkx2drYx5vQfaT8/P1NUVOTss2DBAhMcHGzKysqMMcY88MADpkePHi5j33DDDWb48OHO/br6GVmDQ1VVlYmKijJz5sxxmafdbjevv/66McaYr776ykgyn332mbPPqlWrjM1mM//617+MMca88MILpmXLls45GmPMgw8+aLp27ercv/76683IkSNd6klISDC33XZbrWtxd57GnA5I11577Vlf44vzPHLkiJFkNm3a5BynofyO1qYWd+ZozOk/rL/+A2Tla3M8o2XLlubFF19slOfROkdjGtd5PHHihOncubNZs2aNy7wa67nkFlsd+OWXX7Rjxw4lJyc7j/n5+Sk5OVnZ2dlerOz/7Nu3TzExMerQoYPS0tJUUFAgSdqxY4fKy8tdau/WrZvatm3rrD07O1u9evVy+SDO4cOHy+Fw6Msvv3T2+fUYZ/qcGaM+f0b5+fkqKipyea+QkBAlJCS4zCk0NFQDBgxw9klOTpafn5+2bdvm7DN48GAFBAS4zCk3N1c//fRTreZdm1ou1saNGxUREaGuXbvqjjvu0NGjR51tvjjP4uJiSVJYWJikhvU7Wpta3JnjGa+99prCw8PVs2dPZWRk6OTJk842X5tjZWWlli5dqtLSUiUmJjbK82id4xmN5TxOnTpVI0eOrFZLYzyXUiP5JO2G5scff1RlZWW1T/KOjIzU119/7aWq/k9CQoKysrLUtWtXFRYWatasWbryyiu1Z88eFRUVKSAgoNqX90ZGRqqoqEiSVFRUVOPczrSdq4/D4dDPP/+sn376qd5+Rmdqqum9fl1vRESES7u/v7/CwsJc+rRv377aGGfaWrZsedZ5/3qM89VyMVJSUjRmzBi1b99eeXl5evjhh5Wamqrs7Gw1adLE5+ZZVVWle+65R1dccYV69uzpHLuh/I7WphZ35ihJEyZMULt27RQTE6Pdu3frwQcfVG5urpYvX+5Tc/ziiy+UmJioU6dOqUWLFnr77bfVvXt35eTkNJrzeLY5So3nPC5dulQ7d+7UZ599Vq2tsf03eQYB6RKUmprq/Hd8fLwSEhLUrl07vfnmm2revLkXK8PFGjdunPPfvXr1Unx8vDp27KiNGzcqKSnJi5W5Z+rUqdqzZ4+2bNni7VLqzNnmeOuttzr/3atXL0VHRyspKUl5eXnq2LFjfZfptq5duyonJ0fFxcX6n//5H02cOFGbNm3ydlkedbY5du/evVGcx4MHD2r69Olas2aNmjVr5u1y6g232OpAeHi4mjRpUm3V/OHDhxUVFeWlqs4uNDRUXbp00f79+xUVFaVffvlFx48fd+nz69qjoqJqnNuZtnP1CQ4OVvPmzev1Z3RmvHO9V1RUlI4cOeLSXlFRoWPHjnlk3r9uP18tntShQweFh4dr//79zvf3lXneddddeu+997Rhwwa1adPGebwh/Y7WphZ35liThIQESXI5l74wx4CAAHXq1En9+/dXZmamevfurWeffbZRncezzbEmvnged+zYoSNHjqhfv37y9/eXv7+/Nm3apOeee07+/v6KjIxsNOfy1whIdSAgIED9+/fXunXrnMeqqqq0bt06l/vSDUVJSYny8vIUHR2t/v37q2nTpi615+bmqqCgwFl7YmKivvjiC5c/tGvWrFFwcLDzsnJiYqLLGGf6nBmjPn9G7du3V1RUlMt7ORwObdu2zWVOx48f144dO5x91q9fr6qqKuf/0BITE7V582aVl5e7zKlr165q2bJlreZdm1o86fvvv9fRo0cVHR3tM/M0xuiuu+7S22+/rfXr11e73deQfkdrU4s7c6xJTk6OJLmcy4Y8x7OpqqpSWVlZoziP55tjTXzxPCYlJemLL75QTk6OcxswYIDS0tKc/26U5/KClnSj1pYuXWrsdrvJysoyX331lbn11ltNaGioywp+b7n33nvNxo0bTX5+vvn4449NcnKyCQ8PN0eOHDHGnH5Esm3btmb9+vVm+/btJjEx0SQmJjpff+Zxzauvvtrk5OSY1atXm9atW9f4uOb9999v9u7da+bPn1/j45qe+hmdOHHC7Nq1y+zatctIMnPnzjW7du0y3333nTHm9CPnoaGh5p133jG7d+821157bY2P+fft29ds27bNbNmyxXTu3Nnl8ffjx4+byMhI8+///u9mz549ZunSpSYwMLDa4+/+/v7m6aefNnv37jUzZ86s8fH389XizjxPnDhh7rvvPpOdnW3y8/PN2rVrTb9+/Uznzp3NqVOnfGaed9xxhwkJCTEbN250eTT65MmTzj4N6Xf0fLW4M8f9+/ebxx9/3Gzfvt3k5+ebd955x3To0MEMHjzYZ+ZojDEPPfSQ2bRpk8nPzze7d+82Dz30kLHZbObDDz9sFOfxfHNsLOexJtan8xrDubQiINWh559/3rRt29YEBASYgQMHmq1bt3q7JGPM6ccmo6OjTUBAgPnNb35jbrjhBrN//35n+88//2zuvPNO07JlSxMYGGhGjx5tCgsLXcY4cOCASU1NNc2bNzfh4eHm3nvvNeXl5S59NmzYYPr06WMCAgJMhw4dzOLFi6vV4qmf0YYNG4ykatvEiRONMacfO//rX/9qIiMjjd1uN0lJSSY3N9dljKNHj5rx48ebFi1amODgYHPzzTebEydOuPT5/PPPzaBBg4zdbje/+c1vzOzZs6vV8uabb5ouXbqYgIAA06NHD7Ny5UqX9trU4s48T548aa6++mrTunVr07RpU9OuXTszZcqUaoGzoc+zpvlJcvn9aUi/o7Wp5ULnWFBQYAYPHmzCwsKM3W43nTp1Mvfff7/L5+c09DkaY8wtt9xi2rVrZwICAkzr1q1NUlKSMxzVdlxfnmNjOY81sQakxnAurWzGGHNh15wAAAAaN9YgAQAAWBCQAAAALAhIAAAAFgQkAAAACwISAACABQEJAADAgoAEAABgQUACAACwICABaPA2btwom81W7QsoGxObzaYVK1Z4uwwA/x8BCUCtFBUVadq0aerQoYPsdrtiY2M1atSoal8uebGGDh2qe+65x+XY73//exUWFiokJOSixm4IIeSxxx5Tnz59vFoDgPPz93YBABq+AwcO6IorrlBoaKjmzJmjXr16qby8XB988IGmTp2qr7/+uk7fPyAgQFFRUXX6HgDwa1xBAnBed955p2w2mz799FONHTtWXbp0UY8ePZSenq6tW7c6+xUUFOjaa69VixYtFBwcrOuvv16HDx92tp+5evLf//3fiouLU0hIiMaNG6cTJ05IkiZNmqRNmzbp2Weflc1mk81m04EDB2q8xfZf//Vfio2NVWBgoEaPHq25c+cqNDT0oub54osv6vLLL1ezZs3UrVs3vfDCC862AwcOyGazafny5Ro2bJgCAwPVu3dvZWdnu4xxrrqysrI0a9Ysff755875ZWVlOV/7448/avTo0QoMDFTnzp317rvvXtR8AFyEC/56WwCXlKNHjxqbzWb+9re/nbNfZWWl6dOnjxk0aJDZvn272bp1q+nfv78ZMmSIs8/MmTNNixYtzJgxY8wXX3xhNm/ebKKioszDDz9sjDHm+PHjJjEx0UyZMsUUFhaawsJCU1FRYTZs2GAkmZ9++skYY8yWLVuMn5+fmTNnjsnNzTXz5883YWFhJiQk5Jw1SjJvv/12jW2vvvqqiY6ONm+99Zb59ttvzVtvvWXCwsJMVlaWMcaY/Px8I8l069bNvPfeeyY3N9f88Y9/NO3atXN+I/n56jp58qS59957TY8ePZzzO3nypLO2Nm3amCVLlph9+/aZu+++27Ro0cIcPXr0nHMCUDcISADOadu2bUaSWb58+Tn7ffjhh6ZJkyamoKDAeezLL780ksynn35qjDkdkAIDA43D4XD2uf/++01CQoJzf8iQIWb69OkuY1sD0g033GBGjhzp0ictLe2iAlLHjh3NkiVLXI498cQTJjEx0RjzfwHpxRdfrDa/vXv31rqumTNnmt69e9dY2yOPPOLcLykpMZLMqlWrzjknAHWDW2wAzskYU6t+e/fuVWxsrGJjY53HunfvrtDQUO3du9d5LC4uTkFBQc796OhoHTly5IJqys3N1cCBA12OWfcvRGlpqfLy8jR58mS1aNHCuT355JPKy8tz6RsfH+9SuyRn/Rdb16/HvuyyyxQcHHzBPxsAnsEibQDn1LlzZ9lsNo8txG7atKnLvs1mU1VVlUfGdldJSYmk0+uHEhISXNqaNGnisv/r+m02myR5rP6G+LMBLlVcQQJwTmFhYRo+fLjmz5+v0tLSau1nFk5ffvnlOnjwoA4ePOhs++qrr3T8+HF179691u8XEBCgysrKc/bp2rWrPvvsM5dj1v0LERkZqZiYGH377bfq1KmTy9a+fftaj1ObumozPwDexxUkAOc1f/58XXHFFRo4cKAef/xxxcfHq6KiQmvWrNGCBQu0d+9eJScnq1evXkpLS9O8efNUUVGhO++8U0OGDNGAAQNq/V5xcXHatm2bDhw4oBYtWigsLKxan2nTpmnw4MGaO3euRo0apfXr12vVqlXOKzrnkp+fr5ycHJdjnTt31qxZs3T33XcrJCREKSkpKisr0/bt2/XTTz8pPT29VrXXpq64uDhnDW3atFFQUJDsdnutxgdQf7iCBOC8OnTooJ07d2rYsGG699571bNnT1111VVat26dFixYIOn07aB33nlHLVu21ODBg5WcnKwOHTrojTfeuKD3uu+++9SkSRN1795drVu3VkFBQbU+V1xxhRYuXKi5c+eqd+/eWr16tWbMmKFmzZqdd/z09HT17dvXZdu1a5f+/Oc/68UXX9TixYvVq1cvDRkyRFlZWRd0Bak2dY0dO1YpKSkaNmyYWrdurddff73W4wOoPzZT2xWYANCATZkyRV9//bU++ugjb5fioqHWBeDcuMUGwCc9/fTTuuqqq3TZZZdp1apVevnll10+2JG6AFwMriAB8EnXX3+9Nm7cqBMnTqhDhw6aNm2abr/9dm+X1WDrAnBhCEgAAAAWLNIGAACwICABAABYEJAAAAAsCEgAAAAWBCQAAAALAhIAAIAFAQkAAMCCgAQAAGDx/wC7SZsQBimBCAAAAABJRU5ErkJggg==\n"
          },
          "metadata": {}
        }
      ],
      "source": [
        "# Plot distribution of cotig lengths\n",
        "from Bio import SeqIO\n",
        "from matplotlib import pyplot as plt\n",
        "\n",
        "\n",
        "contig_lens = [len(elem) for elem in SeqIO.parse(\"spade_ecoli/contigs.fasta\", \"fasta\")]\n",
        "plt.hist(contig_lens)\n",
        "plt.xlabel(\"Contig Length\")\n",
        "plt.ylabel(\"Frequency\")\n",
        "plt.show()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "id": "D07WTItLV1wU",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0efcd710-03c3-4e24-e92c-3126173ae496"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Min Contig len : 56\n",
            "Max Contig len : 392317\n",
            "Mean Contig len : 21646.79907\n",
            "Standard Deviation of Contig Len : 47554.72081\n"
          ]
        }
      ],
      "source": [
        "# Report some metrics(min, max, avg, std)\n",
        "import numpy as np\n",
        "\n",
        "stat_d = {\n",
        "    \"Min Contig len\": np.min(contig_lens),\n",
        "    \"Max Contig len\": np.max(contig_lens),\n",
        "    \"Mean Contig len\": np.mean(contig_lens),\n",
        "    \"Standard Deviation of Contig Len\": np.std(contig_lens)\n",
        "}\n",
        "\n",
        "for k, v in stat_d.items():\n",
        "    print(k, \":\", np.round(v, 5))\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Dqv62mp_UWRQ"
      },
      "source": [
        "\n",
        "\n",
        "```\n",
        "# This is formatted as code\n",
        "```\n",
        "\n",
        "- Write your analysis about this work:\n",
        "\n",
        "Most contigs are short, so repetitive areas will be a problem and we will have a more complex de Bruijn graph with possibly more than just one Eulerian path.\n",
        "The standard deviation of contigs is high, which is mostly due to the few contigs of length 350000-400000. Though there aren't many of them and they cannot improve N50 much, they can be key to assembling repetitive areas.\n",
        "\n",
        "The genomes of Ecoli and Citrobacter are 46% similar, which reflects their relative closeness on the tree of species.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JgZJq2u2X2Fu"
      },
      "source": [
        "### Definitions from SPAdes Documentation\n",
        "\n",
        "1. **Chimeric Reads**:\n",
        "A chimeric read or split read means that multiple subsections of that read align to different positions in a reference genome. They are not always a sign of the presence of a PCR chimera and often used to detect structural variations.\n",
        "2. **Bluge**:\n",
        "Bulges and tips occur due to errors in the middle and ends of reads, respectively.\n",
        "3. **k-bimer**:\n",
        "\n",
        " A k-bimer is a triple (a|b, d) consisting of k-mers a and b together with an integer d (estimated distance between particular instances of a and b in a genome).\n",
        "\n",
        " SPAdes first extracts k-bimers from bireads, resulting in k-bimers with inexact distance estimates (inherited from biread distance estimates). The k-bimer adjustment approach transforms this set of k-bimers (with rather inaccurate distance estimates) into a set of adjusted k-bimers with exact or nearly exact distance estimates. Similarly to error correction, which replaces original reads with virtual error-corrected reads, k-bimer adjustment substitutes original k-bimers by adjusted k-bimers.\n",
        "\n",
        "This approach breaks the paired–end reads into pairs of k-mers which are transformed to define pairs of edges (biedges) in the de Bruijn graphs\n",
        "\n",
        "https://en.wikipedia.org/wiki/Chimera_(molecular_biology)#Chimeric_read\n",
        "\n",
        "https://pmc.ncbi.nlm.nih.gov/articles/PMC3342519/\n",
        "\n",
        "https://en.wikipedia.org/wiki/SPAdes_(software)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}